---
title: "Diffferential_peak_calling"
author: "Florian Heigwer"
date: "9/23/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
#library(edgeR)
library(pheatmap)
library(RColorBrewer)
library(ChIPseeker)
library(seqLogo) 
library(rGADEM)
library(rtracklayer)
library(BSgenome.Mmusculus.UCSC.mm10)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(tidyverse)
library(ggplotify)
library(ggpubr)
library(patchwork)
library(cowplot)
library(edgeR)

## B110 Theme

theme_b110<-function(){
  theme_classic() +
  theme(
    axis.text=element_text(size = 10), 
    axis.title=element_text(size = 10),
    plot.title = element_text(size = 12,hjust = 0.5,face="bold"),
    legend.title = element_text(size = 12),
    legend.text = element_text(size =10),
    legend.position = "bottom"
    )
}

## B110 Colors

sgi_blue    = '#5087C8'
sgi_yellow1 = '#F2EE35'
sgi_yellow2 = '#FED98E'
b110_grey   = '#808080'
b110_grey_light   = '#909090'
b110_transparent_black = alpha('#000000',0.5)
google_red='#dd4b39'
google_green='#0F9D58'
google_yellow='#F4B400'
google_blue='#4285F4'

maja_red = alpha('#E64B35CC',0.8)
maja_blue = alpha('#4DBBD5CC',0.8)
maja_young = '#91D1C2CC'
maja_aged = '#3C5488CC'

```

## Raw data read in

Three different 20 month aged mice small intestines were processed together with three different small intestinal samples from 4 month old mice.

All in all, 3 samples of Aged and 3 samples of young mice were subjected to ATAC-seq analysis.

Libraries were processed by Maja Funk.

Sequencing was carried out on an Illumina NextSeq 550 acquiring 75 bp paired end reads using an 150 bp high-output kit by David Ibberson.

Raw Sequence analysis was carried out by Florian heigwer Using The nf-core Atac-seq pipeline version 1.2.1 with default parameters. Based on the ENSEMBL genome release GRCm38.p6.

This results in peak calls for each sample, their nearest genomic feature annotations and according quality controls.

```{r raw_data}

#lets read in the raw peak calls and the read pile ups of the peaks

#in the following we will treat them as read counts from an RNASeq experiment for differential openess testing

raw_peaks <- read_delim(file = "raw_data/consensus_peaks.mLb.clN.featureCounts.txt",delim = "\t",col_types = "ccddcddddddd")
# made using : Program:featureCounts v2.0.1; Command:"featureCounts" "-F" "SAF" "-O" "--fracOverlap" "0.2" "-T" "6" "-p" "--donotsort" "-a" "consensus_peaks.mLb.clN.saf" "-o" "consensus_peaks.mLb.clN.featureCounts.txt" "A_R1.mLb.clN.bam" "Y_R1.mLb.clN.bam" "Y_R2.mLb.clN.bam" "A_R2.mLb.clN.bam" "Y_R3.mLb.clN.bam" "A_R3.mLb.clN.bam" 


#second we load the peak annotations

peak_anno <- read_delim(file ="raw_data/consensus_peaks.mLb.clN.annotatePeaks.txt",delim = "\t")

names(raw_peaks) <- gsub("\\W+","_",names(raw_peaks))
names(peak_anno) <- gsub("\\W+","_",names(peak_anno))

```

# Experimental design

Now we need to define the experimental design matrix accoding to an RNASeq experiment

```{r design matrix}

design_table <- tibble(.rows = 6) 

design_table$sample_names <- raw_peaks %>% dplyr::select(ends_with('bam')) %>% colnames()

design_table$Age <- factor(c('A','Y','Y','A','Y','A'),levels = c('A','Y'))

design_table$Replicate <- factor(c('R1','R1','R2','R2','R3','R3'),levels = c('R1','R2','R3'))

design_table <- design_table  %>% column_to_rownames("sample_names") 

```

# count table objeect creation

```{r deseq_object}
count_data <- raw_peaks %>% 
  column_to_rownames("Geneid") %>% 
  filter(Length>200) %>% 
  dplyr::select(ends_with("bam"))

```

# edgeR pipeline

Because of: 
Gontarz, P., Fu, S., Xing, X. et al. Comparison of differential accessibility analysis strategies for ATAC-seq data. Sci Rep 10, 10150 (2020). https://doi.org/10.1038/s41598-020-66998-4

We are rather using edgeR on this particular data set.

```{r}

edgeR_object <- DGEList(counts=count_data,group = factor(c("A","Y","Y","A","Y","A"))) %>% 
  calcNormFactors() %>%
  estimateCommonDisp() %>% 
  estimateTagwiseDisp()

counts_edgeR <- cpm(edgeR_object)
edgR_stats<- exactTest(edgeR_object,dispersion = "auto",pair = c("Y","A"))

res_edgeR <- 
  edgR_stats %>% 
  topTags(n = Inf,adjust.method = "BH",p.value = Inf) %>% as.data.frame() %>% rownames_to_column("PeakID") %>% as_tibble()

res_edgeR %<>% 
  left_join(peak_anno) %>% 
  dplyr::select(PeakID, log2FoldChange=logFC, padj=FDR, Gene_Name, everything()) %>% 
  mutate(peak_length=End-Start) %>% 
  mutate(stat = qnorm(1-PValue/2,)*sign(log2FoldChange)) %>%
  mutate(stat = if_else(stat==-Inf,-8,if_else(stat==Inf,8,stat)))


txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene

peak <- readPeakFile("raw_data/consensus_peaks.mLb.clN.bed")
seqlevelsStyle(peak) <- "UCSC"

peakAnno.edb <- annotatePeak(peak, tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Mm.eg.db")

plotAnnoPie(peakAnno.edb)

res_edgeR %<>% 
  dplyr::select(PeakID,log2FoldChange,PValue,padj,stat) %>% 
  left_join(peakAnno.edb %>% 
              as.data.frame() %>% 
              as_tibble() %>% 
              rename(PeakID=V4))

write_delim(x = res_edgeR,delim = ",",file = "results/annotated_differential_peaks_edgeR_invivo.csv")

```

We will however use the variance stabilization of DESeq2 to vizualize the the data dispersion.

#simple prefiltering

```{r pre_filter,warning=F,message=F}
deseq_object <- DESeqDataSetFromMatrix(countData = count_data,colData = design_table,design = ~Age)

#excluding genes that were not matched at all
deseq_object <- deseq_object[ rowSums(counts(deseq_object)) > 10, ]


#stabilize variance for low readcount genes

vsd <- vst(deseq_object, blind = FALSE)

#cluster data to get an overview

#Treatments should cluster together

a <- plotPCA(vsd, intgroup = c("Age")) + 
  theme_cowplot()+
  scale_color_manual(values = c(maja_aged,maja_young))

print(a)

ggsave("graphics/PCA_clustering_qc.pdf",a,width=10,height=7,unit="cm")
```

# Volcano of results

```{r volcano}
volcano_data <- res_edgeR %>%
  mutate(signif=if_else(padj<=0.1,if_else(log2FoldChange>0,"up","down"),"none"))

p1 <- volcano_data %>%
  ggplot(aes(x=log2FoldChange,y=-log10(PValue),col=signif,label=SYMBOL))+
    geom_vline(xintercept = 0,lwd=0.5,lty=2,col="grey") +
    geom_point()+
    ggtitle("Getting more or less accessible with age")  +
   scale_color_manual(values = c(maja_blue,b110_grey_light,maja_red)) +
    theme_cowplot() +
    ggrepel::geom_label_repel(data = volcano_data %>% 
                                arrange(padj) %>% 
                                slice_head(n = 20),col="black",min.segment.length = 0) 

p2 <- volcano_data %>%
  ggplot(aes(x=log2FoldChange,y=-log10(PValue),col=signif,label=SYMBOL)) +
    geom_vline(xintercept = 0,lwd=0.5,lty=2, col="grey") +
    geom_point()+
    scale_color_manual(values = c(maja_blue,b110_grey_light,maja_red)) +
    theme_cowplot() +
    ggrepel::geom_label_repel(data = volcano_data %>% 
                                filter(SYMBOL %in% c("Ly6e","Ccr2","Oasl2","Ciita","Dhx58")),min.segment.length = 0,col="black",max.overlaps = 100,size=3)  +
    ylim(0,60) +
    xlim(-4.5,4) +
    theme(legend.position = "none")

print(p1)
print(p2)

ggsave("graphics/Volcano_differentially_open_chromatin_edgeR_focussed_GOI.pdf",p2,width = 8,height = 8,units = "cm")
ggsave("graphics/Volcano_differentially_open_chromatin_edgeR_all.pdf",p1,width = 24,height = 15,units = "cm")

# produce rastered versions
require(ggrastr)

p3 <- volcano_data %>%
  ggplot(aes(x=log2FoldChange,y=-log10(PValue),col=signif,label=SYMBOL))+
    geom_vline(xintercept = 0,lwd=0.5,lty=2,col="grey") +
    geom_point_rast(raster.dpi = 600)+
    ggtitle("Getting more or less accessible with age")  +
   scale_color_manual(values = c(maja_blue,b110_grey_light,maja_red)) +
    theme_cowplot() +
    ggrepel::geom_label_repel(data = volcano_data %>% 
                                arrange(padj) %>% 
                                slice_head(n = 20),col="black",min.segment.length = 0) 

p4 <- volcano_data %>%
  ggplot(aes(x=log2FoldChange,y=-log10(PValue),col=signif,label=SYMBOL)) +
    geom_vline(xintercept = 0,lwd=0.5,lty=2, col="grey") +
    geom_point_rast(raster.dpi = 600)+
    scale_color_manual(values = c(maja_blue,b110_grey_light,maja_red)) +
    theme_cowplot() +
    ggrepel::geom_label_repel(data = volcano_data %>% 
                                filter(SYMBOL %in% c("Ly6e","Ccr2","Oasl2","Ciita","Dhx58")),min.segment.length = 0,col="black",max.overlaps = 100,size=3)  +
    ylim(0,60) +
    xlim(-4.5,4) +
    theme(legend.position = "none")

ggsave("graphics/Volcano_differentially_open_chromatin_edgeR_all_rastered.pdf",p3,width = 24,height = 15,units = "cm")

ggsave("graphics/Volcano_differentially_open_chromatin_edgeR_focussed_GOI_rastered.pdf",p4,width = 8,height = 8,units = "cm")


```

# Example differential peak counts

```{r examples}

topGene <- res_edgeR %>% filter(SYMBOL=="Ccr2") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))

a <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

topGene <- res_edgeR %>% filter(SYMBOL=="Ly6e") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))

b <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")


topGene <- res_edgeR %>% filter(SYMBOL=="Ciita") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
c <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

topGene <- res_edgeR %>% filter(SYMBOL=="Dhx58") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
d <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

topGene <- res_edgeR %>% filter(SYMBOL=="Oasl2") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
e <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

topGene <- res_edgeR %>% filter(SYMBOL=="P2rx7") %>% slice_min(padj) %>% slice_max(abs(log2FoldChange)) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
f <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

g <- (a+b)/(c+d)/(e+f)

print(g)

ggsave("graphics/Example_differentially_accessible_genes.pdf",g,width =15 ,height = 24,units = "cm")

```

# Lets enrich some genesets

Using EdgeR

```{r}

library(fgsea)
library(msigdbr)

# first we get the gene sets from msig db
msig_db_hallmarks<- msigdbr(species = "Mus musculus", category = "H", subcategory = NULL) %>% split(x = .$gene_symbol, f = .$gs_name)
msig_db_go_terms<- msigdbr(species = "Mus musculus", category = "C5", subcategory = "GO:BP") %>% split(x = .$gene_symbol, f = .$gs_name)
msig_db_reactome<- msigdbr(species = "Mus musculus", category = "C2", subcategory = "CP:REACTOME") %>% split(x = .$gene_symbol, f = .$gs_name)

seq_results <- res_edgeR %>%
  group_by(SYMBOL) %>% 
  arrange(padj) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  filter(is.finite(stat)) %>%
  select(stat=stat,Symbol=SYMBOL) %>%
  na.omit() %>% 
  distinct() %>% 
  pull(stat)

names(seq_results) <-  res_edgeR %>%
  group_by(SYMBOL) %>% 
  arrange(padj) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  filter(is.finite(stat)) %>%
  select(stat=stat,Symbol=SYMBOL) %>%
  na.omit() %>% 
  distinct() %>% 
  pull(Symbol)


fgsea_GO <- fgsea(pathways=msig_db_go_terms, stats=seq_results)
 fgsea_GOdy <- fgsea_GO %>%
  as_tibble() %>%
  arrange(padj)

p1 <- fgsea_GOdy %>% 
 arrange(padj) %>%
  filter(padj<0.2) %>%
  head(n=30) %>%
  ggplot(aes( x = NES,y=reorder(pathway, NES),col=padj,size=size)) +
  geom_vline(xintercept = 0,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  geom_point() +
   geom_hline(yintercept = 1:30,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  labs(y="Pathway", x="Normalized Enrichment Score",
       title="GO:BP NES from GSEA") + 
  geom_point() +
  theme_cowplot() +
  scale_color_viridis_c() 

ggsave("graphics/GSEA_GO_BioProcess_edgeR.pdf",p1,width = 44,height = 15,units = "cm")

fgsea_hallmarks <- fgsea(pathways=msig_db_hallmarks, stats=seq_results)
 fgsea_hallmarksdy <- fgsea_hallmarks %>%
  as_tibble() %>%
  arrange(padj)

p1 <- fgsea_hallmarksdy %>% 
 arrange(padj) %>%
  filter(padj<0.2) %>%
  head(n=30) %>%
  ggplot(aes( x = NES,y=reorder(pathway, NES),col=padj,size=size)) +
  geom_vline(xintercept = 0,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  geom_point() +
   geom_hline(yintercept = 1:30,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  labs(y="Pathway", x="Normalized Enrichment Score",
       title="hallmarks:BP NES from GSEA") + 
  geom_point() +
  theme_cowplot() +
  scale_color_viridis_c() 

ggsave("graphics/GSEA_hallmarks_BioProcess_edgeR.pdf",p1,width = 44,height = 15,units = "cm")


fgsea_reactome<- fgsea(pathways=msig_db_reactome, stats=seq_results)
 fgsea_reactomedy <- fgsea_reactome %>%
  as_tibble() %>%
  arrange(padj)

fgsea_reactomedy %>% 
  arrange(padj) %>%
  filter(padj<0.4) %>%
  head(n=30) %>%
  ggplot(aes( x = NES,y=reorder(pathway, NES),col=padj,size=size)) +
  geom_vline(xintercept = 0,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  geom_point() +
   geom_hline(yintercept = 1:30,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  labs(y="Pathway", x="Normalized Enrichment Score",
       title="Reactome NES from GSEA") + 
  geom_point() +
  theme_cowplot() +
  scale_color_viridis_c() 


ggsave("graphics/GSEA_Reactome_Pathways_edgeR.pdf",width = 24,height = 15,units = "cm")


```

# Barcode of enriched BioProcess

```{r}

custom_barcode_plot <- function(df, sig, mlist){
  ## named vector of gene-level stats
  stat_vector <- setNames(df$t, df$Symbol)
  ## genes in signature
  sig_genes <- mlist[[sig]]
  
  ## generate barcode plot
  bc_plot <- plotEnrichment(sig_genes, stat_vector)
  
  ## remove unwanted layers
  bc_plot$layers <- list()
  
  ## add barcode at the bottom
  lowest_pos <- min(bc_plot$data[,2])
  dash_length <- abs(purrr::reduce(range(bc_plot$data[,2]), `-`)*0.1)
  middle <- which.min(abs(sort(df$t, decreasing=T)))
  
  bc_plot_custom <- bc_plot + geom_segment(aes(x=x, xend=x), y=lowest_pos,
                           yend=lowest_pos-dash_length) + 
    geom_line(colour='#4daf4a') + 
    geom_hline(yintercept=lowest_pos, colour='#cccccc') + 
    geom_hline(yintercept=0, colour='#cccccc') + xlab('') +
    theme_classic() +
    geom_tile(data=tibble(rank=1:length(stat_vector), 
                          y=lowest_pos-(1.25*dash_length)), 
              aes(x=rank, y=y, fill=rank),
                  width=1,
                  height=0.5*dash_length) +
    scale_fill_gradient2(low ='#b2182b', high='#2166ac', 
                         mid='#f7f7f7', midpoint = middle) + 
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) +
    theme(panel.grid=element_blank(), 
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          legend.position = 'none') + 
    ggtitle(paste(sig, 'signature')) +
    ylab('Enrichment score')
  
  return(bc_plot_custom)
}

this_data <- res_edgeR %>%
  group_by(SYMBOL) %>% 
  arrange(padj) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  filter(is.finite(stat)) %>%
  select(t=stat,Symbol=SYMBOL) %>%
  na.omit() %>% 
  distinct()

up_fgsea <- fgsea_GO %>% 
 arrange(padj) %>%
  filter(padj<0.2) %>%
  head(n=30)

#bc_plots_up <- map(c(2,7), function(j){#nrow()
  bcp <- custom_barcode_plot(this_data, up_fgsea$pathway[3],msig_db_go_terms) + 
    annotate('text', x=Inf , y=Inf, hjust=1, vjust=1, 
             label=paste('NES =', round(up_fgsea$NES[3], 2), 
                         '\nFDR =', round(up_fgsea$padj[3], 3)))
  
    bcp2 <- custom_barcode_plot(this_data, up_fgsea$pathway[18],msig_db_go_terms) + 
    annotate('text', x=Inf , y=Inf, hjust=1, vjust=1, 
             label=paste('NES =', round(up_fgsea$NES[18], 2), 
                         '\nFDR =', round(up_fgsea$padj[18], 3)))

 ggsave("graphics/GSEA_Reactome_Pathways_edgeR_barcodes.pdf",bcp/bcp2,width = 24,height = 15,units = "cm")
   
    
up_fgsea <- fgsea_hallmarks %>% 
 arrange(padj) %>%
  filter(padj<0.2) %>%
  head(n=30)

#bc_plots_up <- map(c(2,7), function(j){#nrow()
  bcp <- custom_barcode_plot(this_data, up_fgsea$pathway[2],msig_db_hallmarks) + 
    annotate('text', x=Inf , y=Inf, hjust=1, vjust=1, 
             label=paste('NES =', round(up_fgsea$NES[2], 2), 
                         '\nFDR =', round(up_fgsea$padj[2], 3)))
  
    bcp2 <- custom_barcode_plot(this_data, up_fgsea$pathway[3],msig_db_hallmarks) + 
    annotate('text', x=Inf , y=Inf, hjust=1, vjust=1, 
             label=paste('NES =', round(up_fgsea$NES[3], 2), 
                         '\nFDR =', round(up_fgsea$padj[3], 3)))
    
   ggsave("graphics/GSEA_hallmarks_edgeR_barcodes.pdf",bcp/bcp2,width = 24,height = 15,units = "cm")
  
```


# RNA-Seq comparison 

Here we filter for all results that are significant in any of the two compared experimental modalities

```{r}

# load RNA-Seq

rna_data_exvivo <- 
  read_delim(file = "orthogonal_data/bulk_exvivo_all_genes.txt",delim = "\t") %>%
  dplyr::select(SYMBOL=gene_name,log2FoldChange_rna=log2FoldChange,stat_rna=stat,padj_rna=padj,pvalue_rna=pvalue)

#join with ATAC-seq

joined_data_exvivo <- res_edgeR %>%
  dplyr::select(PeakID,log2FoldChange,PValue,padj,SYMBOL) %>%
  left_join(rna_data_exvivo,multiple = "all",by="SYMBOL",relationship = "many-to-many") %>%
  drop_na() %>%
  mutate(log2FoldChange=(log2FoldChange-mean(log2FoldChange))/sd(log2FoldChange),
         log2FoldChange_rna=(log2FoldChange_rna-mean(log2FoldChange_rna))/sd(log2FoldChange_rna))


IFN_genes <- c(msig_db_hallmarks$HALLMARK_INTERFERON_ALPHA_RESPONSE,
               msig_db_hallmarks$HALLMARK_INTERFERON_GAMMA_RESPONSE,
               msig_db_hallmarks$HALLMARK_INFLAMMATORY_RESPONSE,
               msig_db_hallmarks$HALLMARK_IL6_JAK_STAT3_SIGNALING)

TNFA_genes <- msig_db_hallmarks$HALLMARK_TNFA_SIGNALING_VIA_NFKB

ALLO_genes <-msig_db_hallmarks$HALLMARK_ALLOGRAFT_REJECTION

this_data_exvivo <- joined_data_exvivo %>% 
  filter(padj<0.1|padj_rna<0.1) %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
   mutate(TNFG=if_else(SYMBOL %in% TNFA_genes,"TNF","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other"))

p2 <- this_data_exvivo %>%
  ggplot(aes(x=log2FoldChange,y=log2FoldChange_rna,label=SYMBOL,alpha=1-(padj+padj_rna)/2))  +
    geom_smooth(data=this_data_exvivo,method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data_exvivo %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data_exvivo %>% filter(IFNG=="ISG"),col="purple",size=3) +
   geom_point(data = this_data_exvivo %>% filter(TNFG=="TNF"),col=google_yellow,size=3) +
    geom_point(data = this_data_exvivo %>% filter(ALLO=="ALLO"),col=google_green,size=3) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq-invivo")+
    ylab("Log2-FC RNA-Seq-exvivo")

print(p2)

# load RNA-Seq

rna_data_invivo <- 
  read_delim(file = "orthogonal_data/bulk_invivo_all_genes.txt",delim = "\t") %>% 
  dplyr::select(SYMBOL=gene_name,log2FoldChange_rna=log2FoldChange,stat_rna=stat,padj_rna=padj,pvalue_rna=pvalue) %>% 
  mutate(log2FoldChange_rna=-log2FoldChange_rna,stat_rna=-stat_rna)

#join with ATAC-seq

joined_data_invivo <- res_edgeR %>%
  dplyr::select(PeakID,log2FoldChange,PValue,padj,SYMBOL) %>%
  left_join(rna_data_invivo,multiple = "all",by="SYMBOL",relationship = "many-to-many") %>%
  drop_na() %>%
  mutate(log2FoldChange=(log2FoldChange-mean(log2FoldChange))/sd(log2FoldChange),
         log2FoldChange_rna=(log2FoldChange_rna-mean(log2FoldChange_rna))/sd(log2FoldChange_rna))

this_data_invivo <- joined_data_invivo %>% 
  filter(padj<0.1|padj_rna<0.1) %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
   mutate(TNFG=if_else(SYMBOL %in% TNFA_genes,"TNF","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other"))

p3 <- this_data_invivo %>%
  ggplot(aes(x=log2FoldChange,y=log2FoldChange_rna,label=SYMBOL,alpha=1-(padj+padj_rna)/2))  +
    geom_smooth(data=this_data_invivo,method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data_invivo %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data_invivo %>% filter(IFNG=="ISG"),col="purple",size=3) +
   geom_point(data = this_data_invivo %>% filter(TNFG=="TNF"),col=google_yellow,size=3) +
    geom_point(data = this_data_invivo %>% filter(ALLO=="ALLO"),col=google_green,size=3) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq-invivo")+
    ylab("Log2-FC RNA-Seq-invivo")

print(p3)


ggsave("graphics/differntial_comparison_ATAC_RNA_IFNg_ALLO_highlight_exvivo_full.pdf",p2,width = 18,height = 12,units = "cm")

ggsave("graphics/differntial_comparison_ATAC_RNA_IFNg_ALLO_highlight_invivo_full.pdf",p3,width = 18,height = 12,units = "cm")



heat_data <- joined_data_exvivo %>%
  select(SYMBOL,log2FoldChange,log2FoldChange_rna) %>%
  left_join(rna_data_invivo %>% 
              mutate( log2FoldChange_rna=(log2FoldChange_rna-mean(log2FoldChange_rna))/sd(log2FoldChange_rna)) %>%
              rename(log2FoldChange_rna_invivo=log2FoldChange_rna) %>% select(SYMBOL,log2FoldChange_rna_invivo)) %>%
  rename(rna_exvivo=log2FoldChange_rna,rna_invivo=log2FoldChange_rna_invivo,atac_invivo=log2FoldChange) %>%
  filter(abs(atac_invivo)>3) %>% 
  gather(type,value,-SYMBOL) %>%
  distinct() %>%
  filter(SYMBOL %in% c(IFN_genes,ALLO_genes,TNFA_genes))

ordered_names <- heat_data %>% filter(type == "rna_invivo") %>% arrange(value) %>% pull(SYMBOL)


h1 <- ggplot(heat_data,aes(x=type,y=factor(SYMBOL,levels=ordered_names),fill=value)) +
    geom_tile() +
    ylab("gene symbol") +
    scale_fill_gradient2()+
    theme_b110()
  
ggsave("graphics/differntial_comparison_ATAC_RNA_IFNg_ALLO_highlight_invivo_heatmap_rnainsorted_atacfiltered.pdf",h1,width = 12,height = 30,units = "cm")

```

Here we filter for all results that are significant in both of the two compared experimental modalities

```{r}

this_data_exvivo <- joined_data_exvivo %>% 
  filter(padj<0.1,padj_rna<0.1) %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
   mutate(TNFG=if_else(SYMBOL %in% TNFA_genes,"TNF","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other"))

p2 <- this_data_exvivo %>%
  ggplot(aes(x=log2FoldChange,y=log2FoldChange_rna,label=SYMBOL,alpha=1-(padj+padj_rna)/2))  +
    geom_smooth(data=this_data_exvivo,method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data_exvivo %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data_exvivo %>% filter(IFNG=="ISG"),col="purple",size=3) +
   geom_point(data = this_data_exvivo %>% filter(TNFG=="TNF"),col=google_yellow,size=3) +
    geom_point(data = this_data_exvivo %>% filter(ALLO=="ALLO"),col=google_green,size=3) +
   # ggrepel::geom_label_repel(data = this_data %>% filter(IFNG=="ISG"|ALLO=="ALLO"|TNFG=="TNF"),max.overlaps =100) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq")+
    ylab("Log2-FC RNA-Seq-exvivo")

print(p2)

# load RNA-Seq

this_data_invivo <- joined_data_invivo %>% 
  filter(padj<0.1,padj_rna<0.1) %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
   mutate(TNFG=if_else(SYMBOL %in% TNFA_genes,"TNF","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other"))

p3 <- this_data_invivo %>%
  ggplot(aes(x=log2FoldChange,y=log2FoldChange_rna,label=SYMBOL,alpha=1-(padj+padj_rna)/2))  +
    geom_smooth(data=this_data_invivo,method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data_invivo %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data_invivo %>% filter(IFNG=="ISG"),col="purple",size=3) +
   geom_point(data = this_data_invivo %>% filter(TNFG=="TNF"),col=google_yellow,size=3) +
    geom_point(data = this_data_invivo %>% filter(ALLO=="ALLO"),col=google_green,size=3) +
   # ggrepel::geom_label_repel(data = this_data %>% filter(IFNG=="ISG"|ALLO=="ALLO"|TNFG=="TNF"),max.overlaps =100) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq")+
    ylab("Log2-FC RNA-Seq-invivo")

print(p3)


ggsave("graphics/differntial_comparison_ATAC_RNA_IFNg_ALLO_highlight_exvivo_filtered.pdf",p2,width = 18,height = 12,units = "cm")

ggsave("graphics/differntial_comparison_ATAC_RNA_IFNg_ALLO_highlight_invivo_filtered.pdf",p3,width = 18,height = 12,units = "cm")

```

# ATAC-Seq comparison

```{r}

# load ATAC-Seq-exvivo

atac_data_exvivo <- 
  read_delim(file = "orthogonal_data/annotated_differential_peaks_edgeR_exvivo.csv",delim = ",")

ATAC_exvivo_ranges <- makeGRangesFromDataFrame(atac_data_exvivo %>%
                           select(log2FoldChange,padj,start,end,strand,seqnames,SYMBOL),
                         keep.extra.columns=TRUE,
                         ignore.strand=FALSE,
                         seqinfo=NULL,
                         seqnames.field=c("seqnames"),
                         start.field="start",
                         end.field=c("end", "stop"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)

ATAC_invivo_ranges <- makeGRangesFromDataFrame(res_edgeR %>%
                           select(log2FoldChange,padj,start,end,strand,seqnames,SYMBOL),
                         keep.extra.columns=TRUE,
                         ignore.strand=FALSE,
                         seqinfo=NULL,
                         seqnames.field=c("seqnames"),
                         start.field="start",
                         end.field=c("end", "stop"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)

save(ATAC_invivo_ranges,file = "results/ATAC_invivo_ranges.RData")

ATAC_exvivo_ranges_order <- findOverlaps(ATAC_exvivo_ranges,ATAC_invivo_ranges,type="any") %>% queryHits()
ATAC_invivo_ranges_order <- findOverlaps(ATAC_exvivo_ranges,ATAC_invivo_ranges,type="any") %>% subjectHits()

joined_data <- bind_cols(
  atac_data_exvivo[ATAC_exvivo_ranges_order,] %>% 
    dplyr::select(SYMBOL,log2FoldChange_exvivo=log2FoldChange,stat_exvivo=stat,padj_exvivo=padj,pvalue_exvivo=PValue),
  res_edgeR[ATAC_invivo_ranges_order,]  %>%
    dplyr::select(log2FoldChange_invivo=log2FoldChange,stat_invivo=stat,padj_invivo=padj,pvalue_invivo=PValue)
  ) %>%
  drop_na() 

#join with ATAC-seq

IFN_genes <- c(msig_db_hallmarks$HALLMARK_INTERFERON_ALPHA_RESPONSE,
               msig_db_hallmarks$HALLMARK_INTERFERON_GAMMA_RESPONSE,
               msig_db_hallmarks$HALLMARK_INFLAMMATORY_RESPONSE,
               msig_db_hallmarks$HALLMARK_IL6_JAK_STAT3_SIGNALING)

TNFA_genes <- msig_db_hallmarks$HALLMARK_TNFA_SIGNALING_VIA_NFKB

ALLO_genes <-msig_db_hallmarks$HALLMARK_ALLOGRAFT_REJECTION

this_data <- joined_data %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
   mutate(TNFG=if_else(SYMBOL %in% TNFA_genes,"TNF","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other"))

p2 <- this_data %>%
  ggplot(aes(x=log2FoldChange_invivo,y=log2FoldChange_exvivo,label=SYMBOL))  +
    geom_smooth(method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data %>% filter(IFNG=="ISG"),col="#F39B7FFF",size=3) +
   geom_point(data = this_data %>% filter(TNFG=="TNF"),col="#F39B7FFF",size=3) +
    geom_point(data = this_data %>% filter(ALLO=="ALLO"),col="#F39B7FFF",size=3) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq-invivo") +
    ylab("Log2-FC ATAC-Seq-exvivo") + 
  stat_cor(method = "pearson", label.x = 2, label.y = -4)

print(p2)

ggsave("graphics/differntial_comparison_ATAC_IFNg_ALLO_highlight_invivo_vs_exvivo_all.pdf",p2,width = 18,height = 12,units = "cm")

this_data <- joined_data %>% 
   filter(padj_invivo<0.2,padj_exvivo<0.2) %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
   mutate(TNFG=if_else(SYMBOL %in% TNFA_genes,"TNF","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other"))

p2 <- this_data %>%
  ggplot(aes(x=log2FoldChange_invivo,y=log2FoldChange_exvivo,label=SYMBOL))  +
    geom_smooth(method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data %>% filter(IFNG=="ISG"),col="#F39B7FFF",size=3) +
   geom_point(data = this_data %>% filter(TNFG=="TNF"),col="#F39B7FFF",size=3) +
    geom_point(data = this_data %>% filter(ALLO=="ALLO"),col="#F39B7FFF",size=3) +
    ggrepel::geom_label_repel(data = this_data %>% filter(IFNG=="ISG"|ALLO=="ALLO"|TNFG=="TNF"),max.overlaps =100,min.segment.length = 0) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq-invivo")+
    ylab("Log2-FC ATAC-Seq-exvivo")+ 
  stat_cor(method = "spearman", label.x = 2, label.y = -4)

print(p2)

ggsave("graphics/differntial_comparison_ATAC_IFNg_ALLO_highlight_invivo_vs_exvivo_filteredFDR0.2.pdf",p2,width = 18,height = 12,units = "cm")

```

# raw read count comparison

```{r}

raw_peaks_invivo <- read_delim(file = "raw_data/consensus_peaks.mLb.clN.featureCounts.txt",delim = "\t",col_types = "ccddcddddddd")
raw_peaks_exvivo <- read_delim(file = "../ATAC-Seq-exvivo/raw_data/consensus_peaks.mLb.clN.featureCounts.txt",delim = "\t",col_types = "ccddcddddddd")
names(raw_peaks_invivo) <- gsub("\\W+","_",names(raw_peaks_invivo))
names(raw_peaks_exvivo) <- gsub("\\W+","_",names(raw_peaks_exvivo))

ATAC_exvivo_ranges <- makeGRangesFromDataFrame(raw_peaks_exvivo,
                         keep.extra.columns=F,
                         ignore.strand=FALSE,
                         seqinfo=NULL,
                         seqnames.field=c("Chr"),
                         start.field="Start",
                         end.field=c("End", "stop"),
                         strand.field="Strand",
                         starts.in.df.are.0based=FALSE)

ATAC_invivo_ranges <- makeGRangesFromDataFrame(raw_peaks_invivo,
                         keep.extra.columns=F,
                         ignore.strand=FALSE,
                         seqinfo=NULL,
                         seqnames.field=c("Chr"),
                         start.field="Start",
                         end.field=c("End", "stop"),
                         strand.field="Strand",
                         starts.in.df.are.0based=FALSE)

ATAC_exvivo_ranges_order <- findOverlaps(ATAC_exvivo_ranges,ATAC_invivo_ranges,type="any") %>% queryHits()
ATAC_invivo_ranges_order <- findOverlaps(ATAC_exvivo_ranges,ATAC_invivo_ranges,type="any") %>% subjectHits()

joined_data <- bind_cols(
  raw_peaks_invivo[ATAC_invivo_ranges_order,] %>% select(ends_with("bam")),
  raw_peaks_exvivo[ATAC_exvivo_ranges_order,] %>% select(PeakID,ends_with("bam"))
  ) %>%
  drop_na() %>%
  group_by(PeakID) %>%
  summarise_each(sum) %>%
  ungroup()
```

# design table creation

```{r}
design_table <- tibble(.rows = 12) 

design_table$sample_names <- joined_data %>% select(-PeakID) %>% colnames()

design_table$Age <- factor(c('A','Y','Y','A','Y','A','A','Y','A','Y','A','Y'),levels = c('A','Y'))

design_table$Replicate <- factor(c('R1','R1','R2','R2','R3','R3','R3','R3','R1','R2','R2','R1'),levels = c('R1','R2','R3'))

design_table$Origin <- factor(c('invivo','invivo','invivo','invivo','invivo','invivo','exvivo','exvivo','exvivo','exvivo','exvivo','exvivo'),levels = c('invivo','exvivo'))

design_table <- design_table  %>% column_to_rownames("sample_names") 

```

# count table objeect creation

```{r deseq_object}
count_data <- joined_data %>% 
  left_join(raw_peaks_exvivo %>% select(PeakID,Length)) %>%
  filter(Length>200) %>% 
    select(-Length) %>%
  column_to_rownames("PeakID")

deseq_object <- DESeqDataSetFromMatrix(countData = count_data,colData = design_table,design = ~Age+Origin+Age:Origin)

#excluding genes that were not matched at all
deseq_object <- deseq_object[ rowSums(counts(deseq_object)) > 10, ]


#stabilize variance for low readcount genes

vsd <- vst(deseq_object, blind = FALSE)

#cluster data to get an overview
vsd_batchrm <- vsd
assay(vsd_batchrm) <- limma::removeBatchEffect(assay(vsd_batchrm), vsd_batchrm$Origin)

#Treatments should cluster together

a <- plotPCA(vsd, intgroup = c("Age","Origin")) + 
  scale_color_manual(values = c(maja_aged,maja_young,"yellow","green"))+
  theme_classic()+
  ggtitle("raw data")

b <- plotPCA(vsd_batchrm, intgroup = c("Age","Origin")) + 
  scale_color_manual(values = c(maja_aged,maja_young,"yellow","green"))+
  theme_classic() +
  ggtitle("batch corrected")

print(a/b)

ggsave("graphics/PCA_clustering_qc_joined.pdf",a/b,width=10,height=20,unit="cm")


deseq_object <- DESeq(deseq_object)

resultsNames(deseq_object)

res_AY <- results(deseq_object, contrast = c("Age","A","Y")) %>% as.data.frame() %>% rownames_to_column("PeakID") %>% left_join(atac_data_exvivo %>% select(PeakID,SYMBOL))
res_OI <- results(deseq_object, contrast = c("Origin","exvivo","invivo")) %>% as.data.frame() %>% rownames_to_column("PeakID") %>% left_join(atac_data_exvivo %>% select(PeakID,SYMBOL))
res_Int <- results(deseq_object, name = "AgeY.Originexvivo") %>% as.data.frame() %>% rownames_to_column("PeakID") %>% left_join(atac_data_exvivo %>% select(PeakID,SYMBOL))
topGene <- res_Int %>% filter(SYMBOL=="Tmem74") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- 
  plotCounts(deseq_object, gene = topGene, intgroup = c("Age","Origin"),returnData = TRUE) %>% 
    mutate(Age=factor(Age,levels = c("Y","A")),Origin=factor(Origin,levels = c("invivo","exvivo")))

a <- ggplot(geneCounts, aes(x = Age, y = count,col=Origin)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_AY %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))
                    
print(a)

```


#transcription factor binding side analysis

```{r}
require("LOLA")

regionDB = loadRegionDB("orthogonal_data/LOLACore/mm10")

peak <- readPeakFile("raw_data/consensus_peaks.mLb.clN.bed")
seqlevelsStyle(peak) <- "UCSC"

IFN_peak <- 
  res_edgeR %>% 
    filter(SYMBOL %in% msig_db_hallmarks$HALLMARK_INTERFERON_GAMMA_RESPONSE, padj<0.1) %>% pull(PeakID)

locResults = runLOLA(peak[peak$V4 %in% IFN_peak,], peak, regionDB, cores=6)

p1 <- locResults %>% 
  ggplot(aes(x=support,y=pValueLog, label=antibody)) + 
    geom_point() + 
    theme_cowplot() + 
    ggrepel::geom_label_repel(data = locResults %>% arrange(desc(pValueLog)) %>% head(20),max.overlaps=100)

ggsave("graphics/Chip_seek_TFBS_enrichment.pdf",p1,width = 12,height = 16,units = "cm")

print(p1)
```

#ATAC-Seq vizualization

```{r genome_range_setup}
library(Gviz)
library(GenomicRanges)
library(biomaRt)
library(rtracklayer) 
options(ucscChromosomeNames=FALSE)

ma<- useMart('ensembl')

bm <- useMart(host = "http://nov2020.archive.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL",
              dataset = "mmusculus_gene_ensembl")

gtrack <- GenomeAxisTrack()

biomTrack <- BiomartGeneRegionTrack(genome = "GRCm38.p6", name = "Gene structure", 
                                    biomart = bm,
                                    filter = list(with_refseq_mrna=TRUE), 
                                    transcriptAnnotation = "symbol",
                                    stacking = "squish")

scheme <- getScheme()
scheme$BiomartGeneRegionTrack$fill <- "salmon"
scheme$BiomartGeneRegionTrack$col <- NULL
scheme$BiomartGeneRegionTrack$transcriptAnnotation <- "symbol"
addScheme(scheme, "myScheme")
options(Gviz.scheme = "myScheme")

Ain_ranges <- import.bw("raw_data/A.in.mRp.clN.bigWig", as="GRanges") 
Aex_ranges <- import.bw("raw_data/A.ex.mRp.clN.bigWig", as="GRanges") 

Yin_ranges <- import.bw("raw_data/Y.in.mRp.clN.bigWig", as="GRanges")
Yex_ranges <- import.bw("raw_data/Y.ex.mRp.clN.bigWig", as="GRanges") 

Rela_ranges <- import("orthogonal_data/GSM881112_Rela_120_peaks_sorted_lifted_mm10.bed") 
seqlevelsStyle(Rela_ranges) <- "Ensembl"

STAT1_ranges <- import("orthogonal_data/GSM1022317_IFNg_K_STAT1_IFNg_2h_input_peaks_lifted_mm10.bed") 
seqlevelsStyle(STAT1_ranges) <- "Ensembl"

Rela_track <- AnnotationTrack(range = Rela_ranges,
                genome = "GRCm38", 
                data = "score", 
                name = "Rela after LPS",
                col=google_blue,
                fill=google_blue)

STAT1_track <- AnnotationTrack(range = STAT1_ranges,
                genome = "GRCm38", 
                data = "score", 
                name = "STAT1 after IFNg",
                col=google_green,
                fill=google_green)
```

```{r}
range_y <- c(0,1.5)

Yex <- DataTrack(range = Yex_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Young exvivo",
                col=maja_young,
                fill=maja_young,
                ylim = range_y)

Aex <- DataTrack(range = Aex_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Aged exvivo",
                col=maja_aged,
                fill=maja_aged,
                ylim = range_y)


Yin <- DataTrack(range = Yin_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Young invivo",
                col=maja_young,
                fill=maja_young,
                ylim = range_y)

Ain <- DataTrack(range = Ain_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Aged invivo",
                col=maja_aged,
                fill=maja_aged,
                ylim = range_y)


```


```{r Ciita_example}
GOI = "Ciita"
POI =c("Interval_53704")

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)

chrom <- 16


ht <- HighlightTrack(trackList = list(biomTrack,Yex,Aex,Yin,Ain),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0),
                     inBackground=TRUE)

pdf("graphics/Ciita_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = (starts_highlight-500), 
           to = (end_highlight+18000))
dev.off()

```



```{r Usp18_example}
GOI = "Usp18"
POI =c("Interval_120035")

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)

chrom <- 6


ht <- HighlightTrack(trackList = list(biomTrack,Yex,Aex,Yin,Ain),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0),
                     inBackground=TRUE)

pdf("graphics/Usp18_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = (starts_highlight-5000), 
           to = (end_highlight+4000))
dev.off()

```



```{r P2rx7_example}
GOI = "P2rx7"
POI =c("Interval_110724")

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)

chrom <- 5


ht <- HighlightTrack(trackList = list(biomTrack,Yex,Aex,Yin,Ain),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0),
                     inBackground=TRUE)

pdf("graphics/P2rx7_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = (starts_highlight-3000), 
           to = (end_highlight+3000))
dev.off()

```



```{r Dhx58_example}
GOI = "Dhx58"
POI =c("Interval_25527")

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)

chrom <- 11


ht <- HighlightTrack(trackList = list(biomTrack,Yex,Aex,Yin,Ain),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0),
                     inBackground=TRUE)

pdf("graphics/Dhx58_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = (starts_highlight-3000), 
           to = (end_highlight+3000))
dev.off()

```

```{r Ccr2_example}
GOI = "Ccr2"
POI =c("Interval_147305")

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)

chrom <- 9


ht <- HighlightTrack(trackList = list(biomTrack,Yex,Aex,Yin,Ain),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0),
                     inBackground=TRUE)

pdf("graphics/Ccr2_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = (starts_highlight-500), 
           to = (end_highlight+23000))
dev.off()

```

