---
title: "Diffferential_peak_calling"
author: "Florian Heigwer"
date: "9/23/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(edgeR)
library(pheatmap)
library(RColorBrewer)
library(ChIPseeker)
library(ggplotify)
library(patchwork)
library(seqLogo) 
library(rGADEM)
library(rtracklayer)
library(BSgenome.Mmusculus.UCSC.mm10)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(tidyverse)
library(cowplot)
## B110 Theme

theme_b110<-function(){
  theme_classic() +
  theme(
    axis.text=element_text(size = 10), 
    axis.title=element_text(size = 10),
    plot.title = element_text(size = 12,hjust = 0.5,face="bold"),
    legend.title = element_text(size = 12),
    legend.text = element_text(size =10),
    legend.position = "bottom"
    )
}

## B110 Colors

sgi_blue    = '#5087C8'
sgi_yellow1 = '#F2EE35'
sgi_yellow2 = '#FED98E'
b110_grey   = '#808080'
b110_grey_light   = '#909090'
b110_transparent_black = alpha('#000000',0.5)
google_red='#dd4b39'
google_green='#0F9D58'
google_yellow='#F4B400'
google_blue='#4285F4'

maja_red = alpha('#E64B35CC',0.8)
maja_blue = alpha('#4DBBD5CC',0.8)
maja_young = '#91D1C2CC'
maja_aged = '#3C5488CC'

```

## Raw data read in

Three different passages and organoid line combinations for the orgoids from 20 month aged mice sample were processed together with three different passages and organoid line combinations for the 4 month old mice.

All in all, 3 samples from orgoinds of Aged and 3 samples from organoids of young mice were subjected to ATAC-seq analysis.

Libraries were processed by Maja Funk.

Sequencing was carried out on an Illumina NextSeq 550 acquiring 75 bp paired end reads using an 150 bp high-output kit by David Ibberson.

Raw Sequence analysis was carried out by Florian heigwer Using The nf-core Atac-seq pipeline version 1.2.1 with default parameters. Based on the ENSEMBL genome release GRCm38.p6.

This results in peak calls for each sample, their nearest genomic feature annotations and according quality controls.

```{r raw_data}

#lets read in the raw peak calls and the read pile ups of the peaks

#in the following we will treat them as read counts from an RNASeq experiment for differential openess testing

raw_peaks <- read_delim(file = "raw_data/consensus_peaks.mLb.clN.featureCounts.txt",delim = "\t",col_types = "ccddcddddddd")

#second we load the peak annoations

peak_anno <- read_delim(file ="raw_data/consensus_peaks.mLb.clN.annotatePeaks.txt",delim = "\t")

names(raw_peaks) <- gsub("\\W+","_",names(raw_peaks))
names(peak_anno) <- gsub("\\W+","_",names(peak_anno))

```

# Experimental design

Now we need to define the experimental design matrix accoding to an RNASeq experiment

```{r design matrix}

design_table <- tibble(.rows = 6) 

design_table$sample_names <- raw_peaks %>% select(ends_with('bam')) %>% colnames()

design_table$Age <- factor(c('A','Y','A','Y','A','Y'),levels = c('A','Y'))

design_table$Replicate <- factor(c('R3','R3','R1','R2','R2','R1'),levels = c('R1','R2','R3'))

design_table$line <- factor(c('O54','O54','O52','O53','O53','O52'),levels = c('O52','O53','O54'))

design_table$passage <- factor(c(2,2,4,3,3,4))

design_table <- design_table  %>% column_to_rownames("sample_names") 

```

# count table objeect creation

```{r deseq_object}
count_data <- raw_peaks %>% 
  column_to_rownames("PeakID") %>% 
  filter(Length>200) %>% 
  select(ends_with("bam"))

```

# edgeR pipeline

Becasue of: 
Gontarz, P., Fu, S., Xing, X. et al. Comparison of differential accessibility analysis strategies for ATAC-seq data. Sci Rep 10, 10150 (2020). https://doi.org/10.1038/s41598-020-66998-4

We are rather using edgeR on this particular data set.

```{r}

edgeR_object <- DGEList(counts=count_data,group = factor(c("A","Y","A","Y","A","Y"))) %>% 
  calcNormFactors() %>%
  estimateCommonDisp() %>% 
  estimateTagwiseDisp()

counts_edgeR <- cpm(edgeR_object)
edgR_stats<- exactTest(edgeR_object,dispersion = "auto",pair = c("Y","A"))

res_edgeR <- 
  edgR_stats %>% 
  topTags(n = Inf,adjust.method = "BH",p.value = Inf) %>% as.data.frame() %>% rownames_to_column("PeakID") %>% as_tibble()

res_edgeR %<>% 
  left_join(peak_anno) %>% 
  select(PeakID, log2FoldChange=logFC, padj=FDR, Gene_Name, everything()) %>% 
  mutate(peak_length=End-Start) %>% 
  mutate(stat = qnorm(1-PValue/2,)*sign(log2FoldChange)) %>%
  mutate(stat = if_else(stat==-Inf,-8,if_else(stat==Inf,8,stat)))


txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene

peak <- readPeakFile("raw_data/consensus_peaks.mLb.clN.bed")
seqlevelsStyle(peak) <- "UCSC"

peakAnno.edb <- annotatePeak(peak, tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Mm.eg.db")

plotAnnoPie(peakAnno.edb)

res_edgeR %<>% 
  select(PeakID,log2FoldChange,PValue,padj,stat) %>% 
  left_join(peakAnno.edb %>% 
              as.data.frame() %>% 
              as_tibble() %>% 
              rename(PeakID=V4))

write_delim(x = res_edgeR,delim = ",",file = "results/annotated_differential_peaks_edgeR.csv")

```

#simple prefiltering

```{r pre_filter,warning=F,message=F}
deseq_object <- DESeqDataSetFromMatrix(countData = count_data,colData = design_table,design = ~Age)

#excluding genes that were not matched at all
deseq_object <- deseq_object[ rowSums(counts(deseq_object)) > 10, ]


#stabilize variance for low readcount genes

vsd <- vst(deseq_object, blind = FALSE)

#cluster data to get an overview

#Treatments should cluster together

a <- plotPCA(vsd, intgroup = c("Age")) + 
  theme_cowplot()+
  scale_color_manual(values = c(maja_aged,maja_young))

print(a)

ggsave("graphics/PCA_clustering_qc.pdf",a,width=10,height=7,unit="cm")
```

# Volcano of results

```{r volcano}
volcano_data <- res_edgeR %>%
  mutate(signif=if_else(padj<=0.1,if_else(log2FoldChange>0,"up","down"),"none")) %>%
  filter(padj<0.5)

p1 <- volcano_data %>%
  ggplot(aes(x=log2FoldChange,y=-log10(PValue),col=signif,label=SYMBOL)) +
    geom_point()+
    ggtitle("Getting more or less accessible with age")  +
   scale_color_manual(values = c(maja_blue,b110_grey_light,maja_red)) +
    theme_cowplot() +
    ggrepel::geom_label_repel(data = volcano_data %>% 
                                arrange(padj) %>% 
                                slice_head(n = 30),col="black") +
    geom_vline(xintercept = c(-.5,.5),lwd=0.5,lty=2)

ggsave("graphics/Volcano_differentially_open_chromatin_edgeR_all.pdf",p1,width = 24,height = 15,units = "cm")

volcano_data <- res_edgeR %>%
  filter(padj<0.5) %>% 
  mutate(signif=if_else(padj<=0.1,if_else(log2FoldChange>0,"up","down"),"none"))

p2 <- volcano_data %>%
  ggplot(aes(x=log2FoldChange,y=-log10(PValue),col=signif,label=SYMBOL)) +
    geom_point()+
    scale_color_manual(values = c(maja_blue,b110_grey_light,maja_red)) +
    theme_cowplot() +
    ggrepel::geom_label_repel(data = volcano_data %>% 
                                filter(SYMBOL %in% c("Ly6e","Ccr2","Oasl2","Ciita","Dhx58")),min.segment.length = 0,col="black",max.overlaps = 100,size=3) +
    geom_vline(xintercept = c(-.5,.5),lwd=0.5,lty=2) +
    ylim(0,60) +
    xlim(-4.5,4) +
    theme(legend.position = "none")

print(p2)

ggsave("graphics/Volcano_differentially_open_chromatin_edgeR_focussed_GOI.pdf",p2,width = 8,height = 8,units = "cm")


```

# Example differential peak counts

```{r examples}

topGene <- res_edgeR %>% filter(SYMBOL=="Ccr2") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
a <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

topGene <- res_edgeR %>% filter(SYMBOL=="Ly6e") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
b <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")


topGene <- res_edgeR %>% filter(SYMBOL=="Ciita") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
c <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

topGene <- res_edgeR %>% filter(SYMBOL=="Dhx58") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
d <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

topGene <- res_edgeR %>% filter(SYMBOL=="Oasl2") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
e <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

topGene <- res_edgeR %>% filter(SYMBOL=="H2-Aa") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
f <- ggplot(geneCounts, aes(x = Age, y = count)) +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_cowplot() +
  scale_fill_manual(values = c(maja_young,maja_aged))+
    theme(legend.position = "none")

g <- (a+b)/(c+d)/(e+f)

print(g)

ggsave("graphics/Example_differentially_accessible_genes.pdf",g,width =15 ,height = 24,units = "cm")

```

# Lets enrich some genesets

Using EdgeR

```{r}

library(fgsea)
library(msigdbr)

# first we get the gene sets from msig db
msig_db_hallmarks<- msigdbr(species = "Mus musculus", category = "H", subcategory = NULL) %>% split(x = .$gene_symbol, f = .$gs_name)
msig_db_go_terms<- msigdbr(species = "Mus musculus", category = "C5", subcategory = "GO:BP") %>% split(x = .$gene_symbol, f = .$gs_name)
msig_db_reactome<- msigdbr(species = "Mus musculus", category = "C2", subcategory = "CP:REACTOME") %>% split(x = .$gene_symbol, f = .$gs_name)

seq_results <- res_edgeR %>%
  group_by(SYMBOL) %>% 
  arrange(padj) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  filter(is.finite(stat)) %>%
  select(stat=stat,Symbol=SYMBOL) %>%
  na.omit() %>% 
  distinct() %>% 
  pull(stat)

names(seq_results) <-  res_edgeR %>%
  group_by(SYMBOL) %>% 
  arrange(padj) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  filter(is.finite(stat)) %>%
  select(stat=stat,Symbol=SYMBOL) %>%
  na.omit() %>% 
  distinct() %>% 
  pull(Symbol)


fgsea_GO <- fgsea(pathways=msig_db_go_terms, stats=seq_results)
 fgsea_GOdy <- fgsea_GO %>%
  as_tibble() %>%
  arrange(padj)

p1 <- fgsea_GOdy %>% 
 arrange(padj) %>%
  filter(padj<0.2) %>%
  head(n=30) %>%
  ggplot(aes( x = NES,y=reorder(pathway, NES),col=padj,size=size)) +
  geom_vline(xintercept = 0,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  geom_point() +
   geom_hline(yintercept = 1:30,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  labs(y="Pathway", x="Normalized Enrichment Score",
       title="GO:BP NES from GSEA") + 
  geom_point() +
  theme_cowplot() +
  scale_color_viridis_c() 

ggsave("graphics/GSEA_GO_BioProcess_edgeR.pdf",p1,width = 24,height = 15,units = "cm")

fgsea_reactome<- fgsea(pathways=msig_db_reactome, stats=seq_results)
 fgsea_reactomedy <- fgsea_reactome %>%
  as_tibble() %>%
  arrange(padj)

fgsea_reactomedy %>% 
  arrange(padj) %>%
  filter(padj<0.4) %>%
  head(n=30) %>%
  ggplot(aes( x = NES,y=reorder(pathway, NES),col=padj,size=size)) +
  geom_vline(xintercept = 0,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  geom_point() +
   geom_hline(yintercept = 1:30,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  labs(y="Pathway", x="Normalized Enrichment Score",
       title="Reactome NES from GSEA") + 
  geom_point() +
  theme_cowplot() +
  scale_color_viridis_c() 


ggsave("graphics/GSEA_Reactome_Pathways_edgeR.pdf",width = 24,height = 15,units = "cm")


```

# RNA-Seq comparison

```{r}

# load RNA-Seq

rna_data_exvivo <- 
  read_delim(file = "orthogonal_data/bulk_exvivo_all_genes.txt",delim = "\t") %>%
  dplyr::select(SYMBOL=gene_name,log2FoldChange_rna=log2FoldChange,stat_rna=stat,padj_rna=padj,pvalue_rna=pvalue)

#join with ATAC-seq

joined_data <- res_edgeR %>%
  dplyr::select(PeakID,log2FoldChange,PValue,padj,SYMBOL) %>%
  left_join(rna_data_exvivo) %>%
  drop_na() %>%
  mutate(log2FoldChange=(log2FoldChange-mean(log2FoldChange))/sd(log2FoldChange),
         log2FoldChange_rna=(log2FoldChange_rna-mean(log2FoldChange_rna))/sd(log2FoldChange_rna))


IFN_genes <- c(msig_db_hallmarks$HALLMARK_INTERFERON_ALPHA_RESPONSE,
               msig_db_hallmarks$HALLMARK_INTERFERON_GAMMA_RESPONSE,
               msig_db_hallmarks$HALLMARK_INFLAMMATORY_RESPONSE,
               msig_db_hallmarks$HALLMARK_IL6_JAK_STAT3_SIGNALING)

TNFA_genes <- msig_db_hallmarks$HALLMARK_TNFA_SIGNALING_VIA_NFKB

ALLO_genes <-msig_db_hallmarks$HALLMARK_ALLOGRAFT_REJECTION

this_data <- joined_data %>% 
  filter(padj<0.1,padj_rna<0.1) %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
   mutate(TNFG=if_else(SYMBOL %in% TNFA_genes,"TNF","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other"))

p2 <- this_data %>%
  ggplot(aes(x=log2FoldChange,y=log2FoldChange_rna,label=SYMBOL))  +
    geom_smooth(data=this_data,method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data %>% filter(IFNG=="ISG"),col="purple",size=3) +
   geom_point(data = this_data %>% filter(TNFG=="TNF"),col=google_yellow,size=3) +
    geom_point(data = this_data %>% filter(ALLO=="ALLO"),col=google_green,size=3) +
    ggrepel::geom_label_repel(data = this_data %>% filter(IFNG=="ISG"|ALLO=="ALLO"|TNFG=="TNF"),max.overlaps =100) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq")+
    ylab("Log2-FC RNA-Seq-exvivo")

print(p2)

# load RNA-Seq

rna_data_invivo <- 
  read_delim(file = "orthogonal_data/bulk_invivo_all_genes.txt",delim = "\t") %>% 
  dplyr::select(SYMBOL=gene_name,log2FoldChange_rna=log2FoldChange,stat_rna=stat,padj_rna=padj,pvalue_rna=pvalue) %>% 
  mutate(log2FoldChange_rna=-log2FoldChange_rna,stat_rna=-stat_rna)

#join with ATAC-seq

joined_data <- res_edgeR %>%
  dplyr::select(PeakID,log2FoldChange,PValue,padj,SYMBOL) %>%
  left_join(rna_data_invivo) %>%
  drop_na() %>%
  mutate(log2FoldChange=(log2FoldChange-mean(log2FoldChange))/sd(log2FoldChange),
         log2FoldChange_rna=(log2FoldChange_rna-mean(log2FoldChange_rna))/sd(log2FoldChange_rna))

this_data <- joined_data %>% 
  filter(padj<0.1,padj_rna<0.1) %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
   mutate(TNFG=if_else(SYMBOL %in% TNFA_genes,"TNF","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other"))

p3 <- this_data %>%
  ggplot(aes(x=log2FoldChange,y=log2FoldChange_rna,label=SYMBOL))  +
    geom_smooth(data=this_data,method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data %>% filter(IFNG=="ISG"),col="purple",size=3) +
   geom_point(data = this_data %>% filter(TNFG=="TNF"),col=google_yellow,size=3) +
    geom_point(data = this_data %>% filter(ALLO=="ALLO"),col=google_green,size=3) +
    ggrepel::geom_label_repel(data = this_data %>% filter(IFNG=="ISG"|ALLO=="ALLO"|TNFG=="TNF"),max.overlaps =100) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq")+
    ylab("Log2-FC RNA-Seq-invivo")

print(p3)


ggsave("graphics/differntial_comparison_ATAC_RNA_IFNg_ALLO_highlight_exvivo.pdf",p2,width = 18,height = 12,units = "cm")

ggsave("graphics/differntial_comparison_ATAC_RNA_IFNg_ALLO_highlight_invivo.pdf",p3,width = 18,height = 12,units = "cm")

p2 + p3

#    geom_smooth(method = 'lm')
```

#transcription factor binding side analysis

```{r}
require("LOLA")

regionDB = loadRegionDB("orthogonal_data/LOLACore/mm10")

peak <- readPeakFile("raw_data/consensus_peaks.mLb.clN.bed")
seqlevelsStyle(peak) <- "UCSC"

IFN_peak <- 
  res_edgeR %>% 
    filter(SYMBOL %in% msig_db_hallmarks$HALLMARK_INTERFERON_GAMMA_RESPONSE, padj<0.1) %>% pull(PeakID)

locResults = runLOLA(peak[peak$V4 %in% IFN_peak,], peak, regionDB, cores=6)

p1 <- locResults %>% 
  ggplot(aes(x=support,y=pValueLog, label=antibody)) + 
    geom_point() + 
    theme_cowplot() + 
    ggrepel::geom_label_repel(data = locResults %>% arrange(desc(pValueLog)) %>% head(20),max.overlaps=100)

ggsave("graphics/Chip_seek_TFBS_enrichment.pdf",p1,width = 12,height = 16,units = "cm")

```

#ATAC-Seq vizualization

```{r genome_range_setup}
library(Gviz)
library(GenomicRanges)
library(biomaRt)
library(rtracklayer) 
options(ucscChromosomeNames=FALSE)

ma<- useMart('ensembl')

bm <- useMart(host = "http://nov2020.archive.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL",
              dataset = "mmusculus_gene_ensembl")

gtrack <- GenomeAxisTrack()
#itrack <- IdeogramTrack(genome = "mm10",chromosome = "chr16")
biomTrack <- BiomartGeneRegionTrack(genome = "GRCm38.p6", name = "Gene structure", 
                                    biomart = bm,
                                    filter = list(with_refseq_mrna=TRUE), 
                                    transcriptAnnotation = "symbol",
                                    stacking = "squish")

scheme <- getScheme()
scheme$BiomartGeneRegionTrack$fill <- "salmon"
scheme$BiomartGeneRegionTrack$col <- NULL
scheme$BiomartGeneRegionTrack$transcriptAnnotation <- "symbol"
addScheme(scheme, "myScheme")
options(Gviz.scheme = "myScheme")

A1_ranges <- import.bw("raw_data/A_R1.mLb.clN.bigWig", as="GRanges") 
A2_ranges <- import.bw("raw_data/A_R2.mLb.clN.bigWig", as="GRanges") 
A3_ranges <- import.bw("raw_data/A_R3.mLb.clN.bigWig", as="GRanges") 
Y1_ranges <- import.bw("raw_data/Y_R1.mLb.clN.bigWig", as="GRanges")
Y2_ranges <- import.bw("raw_data/Y_R2.mLb.clN.bigWig", as="GRanges") 
Y3_ranges <- import.bw("raw_data/Y_R3.mLb.clN.bigWig", as="GRanges") 

Rela_ranges <- import("orthogonal_data/GSM881112_Rela_120_peaks_sorted_lifted_mm10.bed") 
STAT1_ranges <- import("orthogonal_data/GSM1022317_IFNg_K_STAT1_IFNg_2h_input_peaks_lifted_mm10.bed") 
```

```{r}
range_y <- c(0,1.5)

A1 <- DataTrack(range = A1_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Aged R1",
                col="#3B3B3BCC",
                fill="#3B3B3BCC",
                ylim = range_y)

A2 <- DataTrack(range = A2_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Aged R2",
                col="#3B3B3BCC",
                fill="#3B3B3BCC",
                ylim = range_y)
A3 <- DataTrack(range = A3_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Aged R3",
                col="#3B3B3BCC",
                fill="#3B3B3BCC",
                ylim = range_y)
Y1 <- DataTrack(range = Y1_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Young R1",
                col="#EFC000CC",
                fill="#EFC000CC",
                ylim = range_y)
Y2 <- DataTrack(range = Y2_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Young R2",
                col="#EFC000CC",
                fill="#EFC000CC",
                ylim = range_y)
Y3 <- DataTrack(range = Y3_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h", #histogram
                name = "Young R3",
                col="#EFC000CC",
                fill="#EFC000CC",
                ylim = range_y)
```


```{r Wnt5a_example}
chrom <- 14
start_region <- 28026707
end_region <- 28649405 

GOI = "Wnt5a"

starts_highlight <- res_edgeR %>% filter(SYMBOL==GOI) %>% filter(padj<0.2) %>% pull(start)
end_highlight <- res_edgeR %>% filter(SYMBOL==GOI) %>% filter(padj<0.2) %>% pull(end)

ht <- HighlightTrack(trackList = list(biomTrack,Y1,Y2,Y3,A1,A2,A3),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0.5),
                     inBackground=TRUE)

pdf("graphics/Wnt5a_peak_zoomed.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = 28490000, # if(start_region<min(starts_highlight)){start_region}else{min(starts_highlight)-1000}, 
           to = 28540000)#if(end_region>max(end_highlight)){end_region}else{max(end_highlight)+1000} )
dev.off()

chrom <- 15
start_region <- 74946000
end_region <- 74961000  

GOI = "Ly6e"
POI =c("Interval_51712","Interval_51713")
#starts_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(Start)
#end_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(End)

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)


ht <- HighlightTrack(trackList = list(biomTrack,Y1,Y2,Y3,A1,A2,A3),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0.5),
                     inBackground=TRUE)

pdf("graphics/Ly6e_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = if(start_region<min(starts_highlight)){start_region}else{min(starts_highlight)-1000}, 
           to = if(end_region>max(end_highlight)){end_region}else{max(end_highlight)+1000} )
dev.off()


chrom <- 16
start_region <- 10470000
end_region <- 10530000

GOI = "Ciita"
POI =c("Interval_55659")
#starts_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(Start)
#end_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(End)

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)


ht <- HighlightTrack(trackList = list(biomTrack,Y1,Y2,Y3,A1,A2,A3),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0.5),
                     inBackground=TRUE)

pdf("graphics/Ciita_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = if(start_region<min(starts_highlight)){start_region}else{min(starts_highlight)-1000}, 
           to = if(end_region>max(end_highlight)){end_region}else{max(end_highlight)+1000} )
dev.off()


chrom <- 9
start_region <- 124100000
end_region <- 124113557 

GOI = "Ccr2"
POI =c("Interval_153605")
#starts_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(Start)
#end_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(End)

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)


ht <- HighlightTrack(trackList = list(biomTrack,Y1,Y2,Y3,A1,A2,A3),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.3),
                     fill=alpha(b110_grey_light,0.3),
                     inBackground=TRUE)

pdf("graphics/Ccr2_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = if(start_region<min(starts_highlight)){start_region}else{min(starts_highlight)-1000}, 
           to = if(end_region>max(end_highlight)){end_region}else{max(end_highlight)+1000} )
dev.off()

```



