---
title: "Diffferential_peak_calling"
author: "Florian Heigwer"
date: "9/23/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(edgeR)
library(pheatmap)
library(RColorBrewer)
library(ChIPseeker)
library(ggplotify)
library(patchwork)
library(seqLogo) 
library(rGADEM)
library(rtracklayer)
library(BSgenome.Mmusculus.UCSC.mm10)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(tidyverse)
## B110 Theme

theme_b110<-function(){
  theme_classic() +
  theme(
    axis.text=element_text(size = 10), 
    axis.title=element_text(size = 10),
    plot.title = element_text(size = 12,hjust = 0.5,face="bold"),
    legend.title = element_text(size = 12),
    legend.text = element_text(size =10),
    legend.position = "bottom"
    )
}

## B110 Colors

sgi_blue    = '#5087C8'
sgi_yellow1 = '#F2EE35'
sgi_yellow2 = '#FED98E'
b110_grey   = '#808080'
b110_grey_light   = '#909090'
b110_transparent_black = alpha('#000000',0.5)
google_red='#dd4b39'
google_green='#0F9D58'
google_yellow='#F4B400'
google_blue='#4285F4'

maja_red = alpha('#E64B35CC',0.8)
maja_blue = alpha('#4DBBD5CC',0.8)
maja_young = '#91D1C2CC'
maja_aged = '#3C5488CC'

```

## Raw data read in

Three different passages and organoid line combinations for the orgoids from 20 month aged mice sample were processed together with three different passages and organoid line combinations for the 4 month old mice.

All in all, 3 samples from orgoinds of Aged and 3 samples from organoids of young mice were subjected to ATAC-seq analysis.

Libraries were processed by Maja Funk.

Sequencing was carried out on an Illumina NextSeq 550 acquiring 75 bp paired end reads using an 150 bp high-output kit by David Ibberson.

Raw Sequence analysis was carried out by Florian heigwer Using The nf-core Atac-seq pipeline version 1.2.1 with default parameters. Based on the ENSEMBL genome release GRCm38.p6.

This results in peak calls for each sample, their nearest genomic feature annotations and according quality controls.

```{r raw_data}

#lets read in the raw peak calls and the read pile ups of the peaks

#in the following we will treat them as read counts from an RNASeq experiment for differential openess testing

raw_peaks <- read_delim(file = "raw_data/consensus_peaks.mLb.clN.featureCounts.txt",delim = "\t",col_types = "ccddcddddddd")

#second we load the peak annoations

peak_anno <- read_delim(file ="raw_data/consensus_peaks.mLb.clN.annotatePeaks.txt",delim = "\t")

names(raw_peaks) <- gsub("\\W+","_",names(raw_peaks))
names(peak_anno) <- gsub("\\W+","_",names(peak_anno))

```

# Experimental design

Now we need to define the experimental design matrix accoding to an RNASeq experiment

```{r design matrix}

design_table <- tibble(.rows = 6) 

design_table$sample_names <- raw_peaks %>% select(ends_with('bam')) %>% colnames()

design_table$Age <- factor(c('A','Y','A','Y','A','Y'),levels = c('A','Y'))

design_table$Replicate <- factor(c('R3','R3','R1','R2','R2','R1'),levels = c('R1','R2','R3'))

design_table$line <- factor(c('O54','O54','O52','O53','O53','O52'),levels = c('O52','O53','O54'))

design_table$passage <- factor(c(2,2,4,3,3,4))

design_table <- design_table  %>% column_to_rownames("sample_names") 

```

# DESeq objeect creation

```{r deseq_object}
count_data <- raw_peaks %>% 
  column_to_rownames("PeakID") %>% 
  filter(Length>200) %>% 
  select(ends_with("bam"))

#raw_peaks %>%


deseq_object <- DESeqDataSetFromMatrix(countData = count_data,colData = design_table,design = ~Age)

```

# edgeR pipeline

Becasue of: 
Gontarz, P., Fu, S., Xing, X. et al. Comparison of differential accessibility analysis strategies for ATAC-seq data. Sci Rep 10, 10150 (2020). https://doi.org/10.1038/s41598-020-66998-4

I believe it is more wise to use edgeR on this particular dataset.

```{r}


edgeR_object <- DGEList(counts=count_data,group = factor(c("A","Y","A","Y","A","Y"))) %>% 
  calcNormFactors() %>%
  estimateCommonDisp() %>% 
  estimateTagwiseDisp()

counts_edgeR <- cpm(edgeR_object)
edgR_stats<- exactTest(edgeR_object,dispersion = "auto",pair = c("Y","A"))

res_edgeR <- 
  edgR_stats %>% 
  topTags(n = Inf,adjust.method = "BH",p.value = Inf) %>% as.data.frame() %>% rownames_to_column("PeakID") %>% as_tibble()

res_edgeR %<>% 
  left_join(peak_anno) %>% 
  select(PeakID, log2FoldChange=logFC, padj=FDR, Gene_Name, everything()) %>% 
  mutate(peak_length=End-Start) %>% 
  mutate(stat = qnorm(1-PValue/2,)*sign(log2FoldChange)) %>%
  mutate(stat = if_else(stat==-Inf,-8,if_else(stat==Inf,8,stat)))


txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene

peak <- readPeakFile("raw_data/consensus_peaks.mLb.clN.bed")
seqlevelsStyle(peak) <- "UCSC"

peakAnno.edb <- annotatePeak(peak, tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Mm.eg.db")

plotAnnoPie(peakAnno.edb)

res_edgeR %<>% 
  select(PeakID,log2FoldChange,PValue,padj,stat) %>% 
  left_join(peakAnno.edb %>% 
              as.data.frame() %>% 
              as_tibble() %>% 
              rename(PeakID=V4))

write_delim(x = res_edgeR,delim = ",",file = "results/annotated_differential_peaks_edgeR.csv")

```

#simple prefiltering

```{r pre_filter,warning=F,message=F}

#counting transcripts

nrow(deseq_object)

#excluding genes that were not matched at all
deseq_object <- deseq_object[ rowSums(counts(deseq_object)) > 10, ]

nrow(deseq_object)

#remaining 156703


#stabilize variance for low readcount genes


vsd <- vst(deseq_object, blind = FALSE)

#cluster data to get an overview

#Treatments should cluster together


sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- rownames(design_table)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
annotations = data.frame("Age"=vsd$Age,"Replicate"=vsd$Replicate,"line"=vsd$line)
row.names(annotations)<- rownames(design_table)

c<- as.ggplot(pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors,annotation_row = annotations)) 

# --> nice it clusters the gene expression profile by age

a <- plotPCA(vsd, intgroup = c("Age")) + theme_classic()+
  scale_color_manual(values = c("#3B3B3BCC","#EFC000CC"))

ggsave("graphics/PCA_clustering_qc.pdf",a,width=7,height=5,unit="cm")

b <- plotPCA(vsd, intgroup = c("Replicate","line")) + theme_b110()

(a+b)/c

  # --> also PCA shows that the first driving force to discriminate the samples is the Age of the mice

ggsave("graphics/PCA_clustering_qc.pdf",(a+b)/c,width=20,height=20,unit="cm")
```


# Volcano of results

```{r volcano}
volcano_data <- res_edgeR %>%
  mutate(signif=if_else(padj<=0.1,if_else(log2FoldChange>0,"up","down"),"none")) %>%
  filter(padj<0.5)

volcano_data %>%
  ggplot(aes(x=log2FoldChange,y=-log10(PValue),col=signif,label=SYMBOL)) +
    geom_point()+
    ggtitle("Getting more or less accessible with age")  +
    scale_color_manual(values = c(google_blue,b110_grey_light,google_red)) +
    theme_b110() +
    ggrepel::geom_label_repel(data = volcano_data %>% 
                                arrange(padj) %>% 
                                slice_head(n = 30))

ggsave("graphics/Volcano_differentially_open_chromatin_edgeR.pdf",width = 24,height = 15,units = "cm")

volcano_data <- res_edgeR %>%
  filter(padj<0.5) %>% 
 # filter(grepl("protein",Detailed_Annotation))  %>%
  #group_by(Gene_Name) %>%
  #summarise(padj = mean(padj), log2FoldChange = mean(log2FoldChange),PValue=median(PValue) ) %>%
  mutate(signif=if_else(padj<=0.1,if_else(log2FoldChange>0,"up","down"),"none"))

library(cowplot)
maja_red = alpha('#E64B35CC',0.8)
maja_blue = alpha('#4DBBD5CC',0.8)
p1 <- volcano_data %>%
  ggplot(aes(x=log2FoldChange,y=-log10(PValue),col=signif,label=SYMBOL)) +
    geom_point()+
    ggtitle("Getting more or less accessible with age")  +
    scale_color_manual(values = c(maja_blue,b110_grey_light,maja_red)) +
    theme_cowplot() +
    ggrepel::geom_label_repel(data = volcano_data %>% 
                                filter(SYMBOL %in% c("Ly6e","Ccr2","Ifit3","Tnf","Isg15","Oas3","Itgam","Irgm2","Oas2","Gpc2","Gpc3","Ciita")),min.segment.length = 0,col="black",size=2) 

ggsave("graphics/Volcano_differentially_open_chromatin_edgeR_protein_mean.pdf",p1,width = 13,height = 7,units = "cm")


```


# Example differential peak counts

```{r examples}

topGene <- res_edgeR %>% filter(SYMBOL=="Ccr2") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
a <- ggplot(geneCounts, aes(x = Age, y = count)) +
  scale_y_log10() +  
  #geom_bar(aes(fill=Age),position = "dodge", stat = "summary", fun = "mean") +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_b110() +
  scale_fill_manual(values = c("#EFC000CC","#3B3B3BCC"))

topGene <- res_edgeR %>% filter(SYMBOL=="Ly6e") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
b <- ggplot(geneCounts, aes(x = Age, y = count)) +
  scale_y_log10() +  
  #geom_bar(aes(fill=Age),position = "dodge", stat = "summary", fun = "mean") +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_b110() +
  scale_fill_manual(values = c("#EFC000CC","#3B3B3BCC"))


topGene <- res_edgeR %>% filter(SYMBOL=="Ciita") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
c <- ggplot(geneCounts, aes(x = Age, y = count)) +
  scale_y_log10() +  
  #geom_bar(aes(fill=Age),position = "dodge", stat = "summary", fun = "mean") +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_b110() +
  scale_fill_manual(values = c("#EFC000CC","#3B3B3BCC"))

topGene <- res_edgeR %>% filter(SYMBOL=="H2-Aa") %>% slice_min(padj) %>% pull(PeakID) 
geneCounts <- plotCounts(deseq_object, gene = topGene, intgroup = c("Age"),returnData = TRUE) %>% mutate(Age=factor(Age,levels = c("Y","A")))
d <- ggplot(geneCounts, aes(x = Age, y = count)) +
  scale_y_log10() +  
  #geom_bar(aes(fill=Age),position = "dodge", stat = "summary", fun = "mean") +
  geom_boxplot(aes(fill=Age)) + 
  geom_point(size = 3) + 
  ggtitle(res_edgeR %>% filter(PeakID==topGene) %>% slice_min(padj) %>% unite(SYMBOL,SYMBOL,padj,sep="_") %>% pull(SYMBOL) %>% unique()) +
  theme_b110() +
  scale_fill_manual(values = c("#EFC000CC","#3B3B3BCC"))

e <- (a+b)/(c+d)

print(e)

ggsave("graphics/Example_differentially_accessible_genes.pdf",e,width = 24,height = 15,units = "cm")

```

# Lets enrich some genesets

Using EdgeR

```{r}

library(fgsea)
library(msigdbr)

# first we get the gene sets from msig db
msig_db_hallmarks<- msigdbr(species = "Mus musculus", category = "H", subcategory = NULL) %>% split(x = .$gene_symbol, f = .$gs_name)
msig_db_immune_terms<- msigdbr(species = "Mus musculus", category = "C7", subcategory = "IMMUNESIGDB") %>% split(x = .$gene_symbol, f = .$gs_name)
msig_db_go_terms<- msigdbr(species = "Mus musculus", category = "C5", subcategory = "GO:BP") %>% split(x = .$gene_symbol, f = .$gs_name)
msig_db_reactome<- msigdbr(species = "Mus musculus", category = "C2", subcategory = "CP:REACTOME") %>% split(x = .$gene_symbol, f = .$gs_name)

seq_results <- res_edgeR %>%
  group_by(SYMBOL) %>% 
  arrange(padj) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  filter(is.finite(stat)) %>%
  select(stat=stat,Symbol=SYMBOL) %>%
  na.omit() %>% 
  distinct() %>% 
  pull(stat)

names(seq_results) <-  res_edgeR %>%
  group_by(SYMBOL) %>% 
  arrange(padj) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  filter(is.finite(stat)) %>%
  select(stat=stat,Symbol=SYMBOL) %>%
  na.omit() %>% 
  distinct() %>% 
  pull(Symbol)

fgsea_GO <- fgsea(pathways=msig_db_go_terms, stats=seq_results)
 fgsea_GOdy <- fgsea_GO %>%
  as_tibble() %>%
  arrange(padj)

fgsea_GOdy %>% 
  arrange(padj) %>%
  head(.,30) %>%
  ggplot(aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.1)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="GO:BioProcess NES from GSEA") + 
  theme_b110()+
  scale_fill_manual(values = c(b110_grey,google_blue))

ggsave("graphics/GSEA_GO_BioProcess_edgeR.pdf",width = 24,height = 15,units = "cm")

fgsea_hallmark <- fgsea(pathways=msig_db_hallmarks, stats=seq_results)

fgsea_hallmarktdy <- fgsea_hallmark %>%
  as_tibble() %>%
  arrange(padj)

p1 <- fgsea_hallmarktdy %>% 
  arrange(padj) %>%
  filter(padj<0.2) %>%
  ggplot(aes( x = NES,y=reorder(pathway, NES),col=padj,size=size)) +
  geom_vline(xintercept = 0,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  geom_point() +
   geom_hline(yintercept = 1:14,lty=1,lwd=0.1,col=b110_grey_light,alpha=0.6)+
  labs(y="Pathway", x="Normalized Enrichment Score",
       title="Hallmarks NES from GSEA") + 
  geom_point() +
  theme_cowplot() +
  scale_color_viridis_c() 

ggsave("graphics/GSEA_Hallmarks_edgeR_infigure.pdf",p1,width = 18,height = 15,units = "cm")

fgsea_immu <- fgsea(pathways=msig_db_immune_terms, stats=seq_results)
 fgsea_immudy <- fgsea_immu %>%
  as_tibble() %>%
  arrange(padj)

fgsea_immudy %>% 
  arrange(padj) %>%
  head(.,30) %>%
  ggplot(aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="ImmuneSignatures NES from GSEA") + 
  theme_b110()+
  scale_fill_manual(values = c(b110_grey,google_blue))

ggsave("graphics/GSEA_ImmuneSignatures_edgeR.pdf",width = 24,height = 15,units = "cm")

fgsea_reactome<- fgsea(pathways=msig_db_reactome, stats=seq_results)
 fgsea_reactomedy <- fgsea_reactome %>%
  as_tibble() %>%
  arrange(padj)

fgsea_reactomedy %>% 
  arrange(padj) %>%
  head(.,30) %>%
  ggplot(aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.35)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Reactome Pathways NES from GSEA") + 
  theme_b110()+
  scale_fill_manual(values = c(b110_grey,google_blue))

ggsave("graphics/GSEA_Reactome_Pathways_edgeR.pdf",width = 24,height = 15,units = "cm")


```

# lets show some barcode plot as an example

### GSEA analysis

We perform gene set enrichment analysis using the Broad Institute's [GSEA](http://software.broadinstitute.org/gsea/index.jsp) [@pmid17644558]. An R version of the algorithm is implemented in the `fgsea` algorithm [@Sergushichev060012], which we use for analysis.

We want to visualize the results as a barcode plot. `fgsea` already implements a nice barcode plot, which we cusotomize a bit to adapt it according to our expectations.

Code for the abrcode plot was inspired by [@pmid31097693] . c Benedikt Rauscher

```{r, results='hide', warning=F, message=F}
custom_barcode_plot <- function(df, sig){
  ## named vector of gene-level stats
  stat_vector <- setNames(df$stat, df$SYMBOL)
  ## genes in signature
  sig_genes <- msig_db_hallmarks[[sig]]
  
  ## generate barcode plot
  bc_plot <- plotEnrichment(sig_genes, stat_vector)
  
  ## remove unwanted layers
  bc_plot$layers <- list()
  
  ## add barcode at the bottom
  lowest_pos <- min(bc_plot$data[,2])
  dash_length <- abs(reduce(range(bc_plot$data[,2]), `-`)*0.1)
  middle <- which.min(abs(sort(df$stat, decreasing=T)))
  
  bc_plot_custom <- bc_plot + geom_segment(aes(x=x, xend=x), y=lowest_pos,
                           yend=lowest_pos-dash_length) + 
    geom_line(colour='#4daf4a') + 
    geom_hline(yintercept=lowest_pos, colour='#cccccc') + 
    geom_hline(yintercept=0, colour='#cccccc') + xlab('') +
    theme_classic() +
    geom_tile(data=tibble(rank=1:length(stat_vector), 
                          y=lowest_pos-(1.25*dash_length)), 
              aes(x=rank, y=y, fill=rank),
                  width=1,
                  height=0.5*dash_length) +
    scale_fill_gradient2(low ='#b2182b', high='#2166ac', 
                         mid='#f7f7f7', midpoint = middle) + 
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) +
    theme(panel.grid=element_blank(), 
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          legend.position = 'none') + 
    ggtitle(paste(sig, 'signature')) +
    ylab('Enrichment score')
  
  return(bc_plot_custom)
}
```

barcode plots

```{r}

seq_results <- res_edgeR %>%
  select(stat=stat,SYMBOL) %>%
  na.omit() %>% 
  distinct() %>% 
  group_by(SYMBOL) %>% 
  summarize(stat=mean(stat)) %>%
  ungroup()

down_fgsea <- fgsea_hallmark %>% arrange(NES)

## generate plots
bc_plots_down <- map(1:4, function(j){#nrow()
  bcp <- custom_barcode_plot(seq_results, down_fgsea$pathway[j]) + 
    annotate('text', x=Inf , y=Inf, hjust=1, vjust=1, 
             label=paste('NES =', round(down_fgsea$NES[j], 2), 
                         '\nFDR =', round(down_fgsea$padj[j], 3)))
  
  return(bcp)
})

up_fgsea <- fgsea_hallmark %>% arrange(-NES)

## generate plots
bc_plots_up <- map(1:4, function(j){#nrow()
  bcp <- custom_barcode_plot(seq_results, up_fgsea$pathway[j]) + 
    annotate('text', x=Inf , y=Inf, hjust=1, vjust=1, 
             label=paste('NES =', round(up_fgsea$NES[j], 2), 
                         '\nFDR =', round(up_fgsea$padj[j], 3)))
  
  return(bcp)
})

## plot to canvas
reduce(c(bc_plots_up), `+`) + plot_layout(ncol=2)

  ggsave("graphics/hallmark_signature_enrichment_upregulated.pdf")

## plot to canvas
reduce(c(bc_plots_down), `+`) + plot_layout(ncol=2)

ggsave("graphics/hallmark_signature_enrichment_downregulated.pdf")


```

# RNA-Seq comparison

```{r}

# load RNA-Seq

rna_data <- 
  read_delim(file = "orthogonal_data/top_table_Young_vs_Aged_Wntinddep_upwithage.txt",delim = "\t") %>% 
  dplyr::select(SYMBOL=gene_name,log2FoldChange_rna=log2FoldChange,stat_rna=stat,padj_rna=padj,pvalue_rna=pvalue)

#join with ATAC-seq

joined_data <- res_edgeR %>%
  dplyr::select(PeakID,log2FoldChange,PValue,padj,SYMBOL) %>%
  left_join(rna_data) %>%
  drop_na() %>%
  mutate(log2FoldChange=(log2FoldChange-mean(log2FoldChange))/sd(log2FoldChange),
         log2FoldChange_rna=(log2FoldChange_rna-mean(log2FoldChange_rna))/sd(log2FoldChange_rna))


IFN_genes <- fgsea_hallmark %>% filter(pathway=="HALLMARK_INTERFERON_GAMMA_RESPONSE") %>% pull(leadingEdge) %>% unlist()
ALLO_genes <- fgsea_hallmark %>% filter(pathway=="HALLMARK_ALLOGRAFT_REJECTION") %>% pull(leadingEdge) %>% unlist()
WNT_genes <- fgsea_hallmark %>% filter(pathway=="HALLMARK_WNT_BETA_CATENIN_SIGNALING") %>% pull(leadingEdge) %>% unlist() %>% c(.,"Wnt5a","Ror2")
  
  
this_data <- joined_data %>% 
  filter(padj<0.1,padj_rna<0.1) %>% 
  mutate(IFNG=if_else(SYMBOL %in% IFN_genes,"ISG","other")) %>%
  mutate(ALLO=if_else(SYMBOL %in% ALLO_genes,"ALLO","other")) %>%
  mutate(WNT=if_else(SYMBOL %in% WNT_genes,"WNT","other")) 

p2 <- this_data %>%
  ggplot(aes(x=log2FoldChange,y=log2FoldChange_rna,label=SYMBOL))  +
    geom_smooth(data=this_data,method = "lm",col=b110_grey_light)+ 
    geom_point(data = this_data %>% filter(IFNG=="other"),col=b110_grey) +
    geom_point(data = this_data %>% filter(IFNG=="ISG"),col="purple",size=3) +
    geom_point(data = this_data %>% filter(ALLO=="ALLO"),col=google_green,size=3) +
   geom_point(data = this_data %>% filter(WNT=="WNT"),col=google_yellow,size=3) +
    ggrepel::geom_label_repel(data = this_data %>% filter(IFNG=="ISG"|ALLO=="ALLO"| WNT =="WNT")) +
    theme_cowplot() +
    xlab("Log2-FC ATAC-Seq")+
    ylab("Log2-FC RNA-Seq")

ggsave("graphics/differntial_comparison_ATAC_RNA_IFNg_ALLO_highlight.pdf",p2,width = 16,height = 10,units = "cm")

#    geom_smooth(method = 'lm')
```


```{r}

data_to_plot <- joined_data %>% 
  filter(SYMBOL %in% IFN_genes) %>% 
  group_by(SYMBOL) %>% 
  arrange(padj) %>%
  slice_head(n=1) %>%
  ungroup() %>%
  dplyr::select(SYMBOL,log2FoldChange,padj,padj_rna,log2FoldChange_rna,pvalue=PValue,pvalue_rna) %>% 
  gather(type,value,-SYMBOL) %>% 
  separate(type,c("type","modality")) %>% 
  mutate(modality=if_else(is.na(modality),"ATAC",modality)) %>% 
  spread(type,value) %>%
  group_by(modality) %>%
  mutate(padj=-log10(padj)) #,padj=(padj-min(padj))/max(padj)

data_to_plot %>% 
  ggplot(aes(x=modality,y=SYMBOL,colour=log2FoldChange,size=padj)) +
    geom_point() +
    scale_color_gradient2(low = "blue",high = "red") +
    theme_b110()
```

#transcription factor binding side analysis

```{r}
require("LOLA")

regionDB = loadRegionDB("orthogonal_data/LOLACore/mm10")

peak <- readPeakFile("raw_data/consensus_peaks.mLb.clN.bed")
seqlevelsStyle(peak) <- "UCSC"

IFN_peak <- res_edgeR %>% filter(SYMBOL %in% IFN_genes, padj<0.1) %>% pull(PeakID)

locResults2 = runLOLA(peak[peak$V4 %in% IFN_peak,], peak, regionDB, cores=6)
```

#ATAC-Seq vizualization

```{r genome_range_setup}
library(Gviz)
library(GenomicRanges)
library(biomaRt)
library(rtracklayer) 
options(ucscChromosomeNames=FALSE)

ma<- useMart('ensembl')

bm <- useMart(host = "http://nov2020.archive.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL",
              dataset = "mmusculus_gene_ensembl")

gtrack <- GenomeAxisTrack()
#itrack <- IdeogramTrack(genome = "mm10",chromosome = "chr16")
biomTrack <- BiomartGeneRegionTrack(genome = "GRCm38.p6", name = "Gene structure", 
                                    biomart = bm,
                                    filter = list(with_refseq_mrna=TRUE), 
                                    transcriptAnnotation = "symbol",
                                    stacking = "squish")

scheme <- getScheme()
scheme$BiomartGeneRegionTrack$fill <- "salmon"
scheme$BiomartGeneRegionTrack$col <- NULL
scheme$BiomartGeneRegionTrack$transcriptAnnotation <- "symbol"
addScheme(scheme, "myScheme")
options(Gviz.scheme = "myScheme")

A1_ranges <- import.bw("raw_data/A_R1.mLb.clN.bigWig", as="GRanges") 
A2_ranges <- import.bw("raw_data/A_R2.mLb.clN.bigWig", as="GRanges") 
A3_ranges <- import.bw("raw_data/A_R3.mLb.clN.bigWig", as="GRanges") 
Y1_ranges <- import.bw("raw_data/Y_R1.mLb.clN.bigWig", as="GRanges")
Y2_ranges <- import.bw("raw_data/Y_R2.mLb.clN.bigWig", as="GRanges") 
Y3_ranges <- import.bw("raw_data/Y_R3.mLb.clN.bigWig", as="GRanges") 

```

```{r}
range_y <- c(0,1.5)

A1 <- DataTrack(range = A1_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Aged R1",
                col="#3B3B3BCC",
                fill="#3B3B3BCC",
                ylim = range_y)

A2 <- DataTrack(range = A2_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Aged R2",
                col="#3B3B3BCC",
                fill="#3B3B3BCC",
                ylim = range_y)
A3 <- DataTrack(range = A3_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Aged R3",
                col="#3B3B3BCC",
                fill="#3B3B3BCC",
                ylim = range_y)
Y1 <- DataTrack(range = Y1_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Young R1",
                col="#EFC000CC",
                fill="#EFC000CC",
                ylim = range_y)
Y2 <- DataTrack(range = Y2_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h",
                name = "Young R2",
                col="#EFC000CC",
                fill="#EFC000CC",
                ylim = range_y)
Y3 <- DataTrack(range = Y3_ranges,
                genome = "GRCm38", 
                data = "score", 
                type = "h", #histogram
                name = "Young R3",
                col="#EFC000CC",
                fill="#EFC000CC",
                ylim = range_y)
```


```{r Wnt5a_example}
chrom <- 14
start_region <- 28026707
end_region <- 28649405 

GOI = "Wnt5a"

starts_highlight <- res_edgeR %>% filter(SYMBOL==GOI) %>% filter(padj<0.2) %>% pull(start)
end_highlight <- res_edgeR %>% filter(SYMBOL==GOI) %>% filter(padj<0.2) %>% pull(end)

ht <- HighlightTrack(trackList = list(biomTrack,Y1,Y2,Y3,A1,A2,A3),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0.5),
                     inBackground=TRUE)

pdf("graphics/Wnt5a_peak_zoomed.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = 28490000, # if(start_region<min(starts_highlight)){start_region}else{min(starts_highlight)-1000}, 
           to = 28540000)#if(end_region>max(end_highlight)){end_region}else{max(end_highlight)+1000} )
dev.off()

chrom <- 15
start_region <- 74946000
end_region <- 74961000  

GOI = "Ly6e"
POI =c("Interval_51712","Interval_51713")
#starts_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(Start)
#end_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(End)

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)


ht <- HighlightTrack(trackList = list(biomTrack,Y1,Y2,Y3,A1,A2,A3),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0.5),
                     inBackground=TRUE)

pdf("graphics/Ly6e_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = if(start_region<min(starts_highlight)){start_region}else{min(starts_highlight)-1000}, 
           to = if(end_region>max(end_highlight)){end_region}else{max(end_highlight)+1000} )
dev.off()


chrom <- 16
start_region <- 10470000
end_region <- 10530000

GOI = "Ciita"
POI =c("Interval_55659")
#starts_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(Start)
#end_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(End)

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)


ht <- HighlightTrack(trackList = list(biomTrack,Y1,Y2,Y3,A1,A2,A3),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.5),
                     fill=alpha(b110_grey_light,0.5),
                     inBackground=TRUE)

pdf("graphics/Ciita_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = if(start_region<min(starts_highlight)){start_region}else{min(starts_highlight)-1000}, 
           to = if(end_region>max(end_highlight)){end_region}else{max(end_highlight)+1000} )
dev.off()


chrom <- 9
start_region <- 124100000
end_region <- 124113557 

GOI = "Ccr2"
POI =c("Interval_153605")
#starts_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(Start)
#end_highlight <- res_edgeR %>% filter(Gene_Name==GOI) %>% filter(padj<0.2) %>% pull(End)

starts_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(start)
end_highlight <- res_edgeR %>% filter(PeakID %in% POI) %>% pull(end)


ht <- HighlightTrack(trackList = list(biomTrack,Y1,Y2,Y3,A1,A2,A3),
                     start = starts_highlight, end = end_highlight,
                     chromosome = chrom,
                     col=alpha(b110_grey_light,0.3),
                     fill=alpha(b110_grey_light,0.3),
                     inBackground=TRUE)

pdf("graphics/Ccr2_peak_large.pdf")
plotTracks(c(gtrack,ht),chromosome = chrom, 
           from = if(start_region<min(starts_highlight)){start_region}else{min(starts_highlight)-1000}, 
           to = if(end_region>max(end_highlight)){end_region}else{max(end_highlight)+1000} )
dev.off()

```



