---
title: "RNA_Seq_Analysis_MJ_YoungvsAge_Bulk_invivo_exvivo"
output:
  html_document: default
  pdf_document: default
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tximportData)
library(ggrepel)
library(DESeq2)
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(cowplot)

```


## B110 Theme

```{r theme, include=FALSE}

theme_b110 <- function(){
  theme_classic() +
  theme(
    axis.text = element_text(size = 10), 
    axis.title = element_text(size = 10),
    plot.title = element_text(size = 12,hjust = 0.5,face = "bold"),
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 10),
    legend.position = "bottom"
    )
  
  theme_cowplot()
}

```

## B110 Colors

```{r colors}

sgi_blue    = '#5087C8'
sgi_yellow1 = '#F2EE35'
sgi_yellow2 = '#FED98E'
b110_grey   = '#808080'
b110_grey_light   = '#909090'
b110_transparent_black = alpha('#000000',0.5)
google_red = '#dd4b39'
google_green = '#0F9D58'
google_yellow = '#F4B400'
google_blue = '#4285F4'
maja_red = alpha('#E64B35CC',0.8)
maja_blue = alpha('#4DBBD5CC',0.8)
maja_young = '#91D1C2CC'
maja_youngD = '#00A087CC'
maja_aged = '#8491B4CC'
maja_agedD = '#3C5488CC'
maja_young2 = '#EFC000CC'
maja_aged2 = '#003C67CC'

```

## Read in the organoid meta data and exclude Wnt data

```{r meta_data}

raw_meta_data1 <- read_delim("./meta_data/15607_meta.tsv",delim = "\t")

raw_meta_data1 %<>% 
  filter(grepl("R1",FASTQ_FILE)) %>% 
  mutate(run = gsub(pattern = "(.+)_.*.fastq.gz",replacement = "\\1",x = FASTQ_FILE)) %>% 
  dplyr::rename(sample = SAMPLE_ID) %>% mutate(batch = "b15607", type = "exvivo")


raw_meta_data2 <- read_delim("./meta_data/16694_meta.tsv",delim = "\t")

raw_meta_data2 %<>% 
  filter(grepl("R1",FASTQ_FILE)) %>% 
  mutate(run = gsub(pattern = "(.+)_.*.fastq.gz",replacement = "\\1",x = FASTQ_FILE)) %>% 
  dplyr::rename(sample = SAMPLE_ID) %>% mutate(batch = "b16694", type = "exvivo")

raw_meta_data <- bind_rows(list(raw_meta_data1,raw_meta_data2))

rm(raw_meta_data1,raw_meta_data2)

raw_meta_data <- raw_meta_data %>% 
  filter(!grepl("Wnt", sample))

```

## Read in the in vivo meta data 

```{r meta_data}

raw_meta_data3 <- read_delim("./meta_data/19397_meta.tsv",delim = "\t")

raw_meta_data3 %<>% 
  filter(grepl("R1",FASTQ_FILE)) %>% 
  mutate(run = gsub(pattern = "(.+)_.*.fastq.gz",replacement = "\\1",x = FASTQ_FILE)) %>% 
  dplyr::rename(sample = SAMPLE_NAME) %>% mutate(batch = "b19397", type = "invivo")

raw_meta_data <- bind_rows(list(raw_meta_data,raw_meta_data3))

rm(raw_meta_data3)
```

#import the data from star_organoids

```{r,warning=F,message=F}

star_data <- read_delim("./raw_data/star_merged_gene_counts.tsv",delim = "\t") %>% 
  select(Geneid, gene_name, starts_with(raw_meta_data$run))

colnames(star_data) <- gsub(pattern = "(\\S+?)Aligned.sortedByCoord.out.bam",replacement = "\\1",x = colnames(star_data))

star_data_id2name <- star_data %>% select(-starts_with("AS"))

star_data %<>% select(-gene_name) %>% column_to_rownames("Geneid") %>% as.data.frame()
```

#import the data from star_invivo

```{r,warning=F,message=F}

#require(tximport)
#txi <- tximport(files, type = "salmon", tx2gene = tx2gene,ignoreTxVersion = TRUE)
#names(txi)

merged_gene_counts = read_delim("./raw_data/merged_gene_counts.txt",delim = "\ ")

star_data2 <- merged_gene_counts  %>% 
  select(-starts_with("Undetermined"))

colnames(star_data2) <- gsub(pattern = "(\\S+?)Aligned.sortedByCoord.out.bam",replacement = "\\1",x = colnames(star_data2))

star_data2_id2name <- star_data2 %>% select(-starts_with("AS"))

star_data2 %<>% select(-gene_name) %>% column_to_rownames("Geneid") %>% as.data.frame()


star_data = cbind.data.frame(star_data, star_data2)
```

#prepare the DESeq2 data object

```{r,warning=F,message=F}
require("DESeq2")

sampleTable<-
  raw_meta_data %>% 
  dplyr::select(sample,run,batch,type) %>% 
  mutate(Age=factor(if_else(grepl("Aged|-A",sample),"Aged","Young"))) %>%
  as.data.frame()

rownames(sampleTable)<-sampleTable$run



```

#simple prefiltering

```{r,warning=F,message=F}

sampleTable <- sampleTable[colnames(star_data),]
dds_star <- DESeqDataSetFromMatrix(star_data, sampleTable, ~type+Age+type:Age)

dds <- dds_star

#counting transcripts

nrow(dds)

#excluding genes that were not matched at all
dds <- dds[ rowSums(counts(dds)) > 1, ]
nrow(dds)

#remaining 26406

```

#stabilize variance for low readcount genes

```{r,warning=F,message=F}

vsd <- vst(dds, blind = FALSE)

```

#cluster data to get an overview

Treatments should cluster together

```{r,warning=F,message=F}
require("pheatmap")
require("RColorBrewer")
require("patchwork")

sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- vsd$sample
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
annotations = data.frame("Age"=vsd$Age,"Source"=vsd$type)
row.names(annotations)<-vsd$sample

my_colour = list(
    Age = c(Aged = maja_agedD, Young = maja_young)
)

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors,annotation_row = annotations, annotation_colors = my_colour)


# --> nice it clusters the gene expression profile by age


a <- plotPCA(vsd, "batch") + ggtitle("PCA by batch")
b <- plotPCA(vsd, "Age") + ggtitle("PCA by Age")
c <- plotPCA(vsd, "type") + ggtitle("PCA by Source")
d <- a+b+c+plot_layout(ncol = 3)

ggsave(d,filename = "PCA_pre_batch_correct.pdf", width = 40, height = 30, units = "cm")

```

# lets visualize the data with the batch effect removed

```{r}
require(patchwork)

  # --> also PCA shows that the first driving force to discriminate the samples is the Age of the mice
vsd_batchrm <- vsd
assay(vsd_batchrm) <- limma::removeBatchEffect(assay(vsd_batchrm), vsd_batchrm$type)

b <- plotPCA(vsd_batchrm, "Age") + ggtitle("PCA by Age") + scale_color_manual(values=c(maja_agedD,maja_youngD)) + theme_cowplot()

b

sampleDists <- dist(t(assay(vsd_batchrm)))
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- vsd$sample
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
annotations = data.frame("Age"=vsd_batchrm$Age,"Source"=vsd_batchrm$type)
row.names(annotations)<-vsd_batchrm$sample

my_colour = list(
    Age = c(Aged = maja_agedD, Young = maja_young)
)

colors <- colorRampPalette( rev(brewer.pal(9, "GnBu")) )(255)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors,annotation_row = annotations, annotation_colors = my_colour)


a <- plotPCA(vsd_batchrm, "type") + ggtitle("PCA by Source") + theme_cowplot()
b <- plotPCA(vsd_batchrm, "Age") + ggtitle("PCA by Age") + scale_color_manual(values=c(maja_agedD,maja_young)) + theme_cowplot()

ggsave("./typebatch_PCA.pdf", width = 5, height = 4)
dev.off()

b

d<-a+b

ggsave(d,filename = "PCA_post_batch_correct.pdf")

```

